<!DOCTYPE HTML>
<html lang="en" class="sidebar-visible no-js light">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>A collection of resources for data scientists (not only) in bioinformatics</title>
        <meta name="robots" content="noindex" />


        <!-- Custom HTML head -->
        
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff" />

        <link rel="icon" href="favicon.svg">
        <link rel="shortcut icon" href="favicon.png">
        <link rel="stylesheet" href="css/variables.css">
        <link rel="stylesheet" href="css/general.css">
        <link rel="stylesheet" href="css/chrome.css">
        <link rel="stylesheet" href="css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="highlight.css">
        <link rel="stylesheet" href="tomorrow-night.css">
        <link rel="stylesheet" href="ayu-highlight.css">

        <!-- Custom theme stylesheets -->

    </head>
    <body>
    <div id="body-container">
        <!-- Provide site root to javascript -->
        <script>
            var path_to_root = "";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('no-js')
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add('js');
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            var html = document.querySelector('html');
            var sidebar = null;
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded affix "><a href="index.html">Introduction</a></li><li class="chapter-item expanded "><a href="bioinformatics/index.html"><strong aria-hidden="true">1.</strong> Bioinformatics</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="bioinformatics/bioconda.html"><strong aria-hidden="true">1.1.</strong> bioconda</a></li><li class="chapter-item expanded "><a href="bioinformatics/bioconductor.html"><strong aria-hidden="true">1.2.</strong> bioconductor</a></li></ol></li><li class="chapter-item expanded "><a href="coding_tools/index.html"><strong aria-hidden="true">2.</strong> Coding tools</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="coding_tools/conventions.html"><strong aria-hidden="true">2.1.</strong> coding conventions</a></li><li class="chapter-item expanded "><a href="coding_tools/ide.html"><strong aria-hidden="true">2.2.</strong> integrated development environments (IDEs)</a></li><li class="chapter-item expanded "><a href="coding_tools/licensing.html"><strong aria-hidden="true">2.3.</strong> licensing</a></li><li class="chapter-item expanded "><a href="coding_tools/markdown.html"><strong aria-hidden="true">2.4.</strong> markdown syntax</a></li><li class="chapter-item expanded "><a href="coding_tools/regular_expressions.html"><strong aria-hidden="true">2.5.</strong> regular expressions</a></li><li class="chapter-item expanded "><a href="coding_tools/user_experience.html"><strong aria-hidden="true">2.6.</strong> user experience</a></li><li class="chapter-item expanded "><a href="coding_tools/version_control.html"><strong aria-hidden="true">2.7.</strong> version control</a></li></ol></li><li class="chapter-item expanded "><a href="command_line/index.html"><strong aria-hidden="true">3.</strong> Command line</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="command_line/bash.html"><strong aria-hidden="true">3.1.</strong> bash</a></li><li class="chapter-item expanded "><a href="command_line/conda.html"><strong aria-hidden="true">3.2.</strong> conda (software installation)</a></li><li class="chapter-item expanded "><a href="command_line/linux_tools.html"><strong aria-hidden="true">3.3.</strong> linux tools</a></li></ol></li><li class="chapter-item expanded "><a href="data_analyses/index.html"><strong aria-hidden="true">4.</strong> Data analyses</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="data_analyses/differential_expression.html"><strong aria-hidden="true">4.1.</strong> differential expression analysis</a></li></ol></li><li class="chapter-item expanded "><a href="data_downloads/index.html"><strong aria-hidden="true">5.</strong> Data downloads</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="data_downloads/ncbi_datasets_cli.html"><strong aria-hidden="true">5.1.</strong> ncbi_datasets</a></li></ol></li><li class="chapter-item expanded "><a href="data_formats/index.html"><strong aria-hidden="true">6.</strong> Data formats and tools</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="data_formats/fastq.html"><strong aria-hidden="true">6.1.</strong> fastq: sequencing data</a></li><li class="chapter-item expanded "><a href="data_formats/tsv.html"><strong aria-hidden="true">6.2.</strong> tab-/comma-separated values</a></li><li class="chapter-item expanded "><a href="data_formats/yaml.html"><strong aria-hidden="true">6.3.</strong> YAML syntax</a></li></ol></li><li class="chapter-item expanded "><a href="data_presentation/index.html"><strong aria-hidden="true">7.</strong> Data presentation</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="data_presentation/color_palettes.html"><strong aria-hidden="true">7.1.</strong> color palettes</a></li><li class="chapter-item expanded "><a href="data_presentation/grammar_of_graphics.html"><strong aria-hidden="true">7.2.</strong> grammar of graphics</a></li><li class="chapter-item expanded "><a href="coding_tools/interactive_reporting.html"><strong aria-hidden="true">7.3.</strong> interactive reporting</a></li></ol></li><li class="chapter-item expanded "><a href="languages/index.html"><strong aria-hidden="true">8.</strong> Programming Languages</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="languages/python.html"><strong aria-hidden="true">8.1.</strong> python</a></li><li class="chapter-item expanded "><a href="languages/r.html"><strong aria-hidden="true">8.2.</strong> R</a></li><li class="chapter-item expanded "><a href="languages/rust.html"><strong aria-hidden="true">8.3.</strong> Rust</a></li></ol></li><li class="chapter-item expanded "><a href="statistics/index.html"><strong aria-hidden="true">9.</strong> Statistics</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="statistics/bayesian.html"><strong aria-hidden="true">9.1.</strong> bayesian statistics (priors, likelihoods and posterior distributions)</a></li><li class="chapter-item expanded "><a href="statistics/frequentist.html"><strong aria-hidden="true">9.2.</strong> frequentist statistics (hypothesis tests, p-values, etc.)</a></li></ol></li><li class="chapter-item expanded "><a href="workflows/index.html"><strong aria-hidden="true">10.</strong> Workflow management</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="workflows/snakemake.html"><strong aria-hidden="true">10.1.</strong> Snakemake</a></li></ol></li><li class="chapter-item expanded "><a href="servers/index.html"><strong aria-hidden="true">11.</strong> Working on servers</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="servers/screen.html"><strong aria-hidden="true">11.1.</strong> screen: persistent shell sessions</a></li><li class="chapter-item expanded "><a href="servers/ssh.html"><strong aria-hidden="true">11.2.</strong> ssh login</a></li><li class="chapter-item expanded "><a href="servers/tmux.html"><strong aria-hidden="true">11.3.</strong> tmux</a></li></ol></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle"></div>
        </nav>

        <!-- Track and set sidebar scroll position -->
        <script>
            var sidebarScrollbox = document.querySelector('#sidebar .sidebar-scrollbox');
            sidebarScrollbox.addEventListener('click', function(e) {
                if (e.target.tagName === 'A') {
                    sessionStorage.setItem('sidebar-scroll', sidebarScrollbox.scrollTop);
                }
            }, { passive: true });
            var sidebarScrollTop = sessionStorage.getItem('sidebar-scroll');
            sessionStorage.removeItem('sidebar-scroll');
            if (sidebarScrollTop) {
                // preserve sidebar scroll position when navigating via links within sidebar
                sidebarScrollbox.scrollTop = sidebarScrollTop;
            } else {
                // scroll sidebar to current active section when navigating via "next/previous chapter" buttons
                var activeSection = document.querySelector('#sidebar .active');
                if (activeSection) {
                    activeSection.scrollIntoView({ block: 'center' });
                }
            }
        </script>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky">
                    <div class="left-buttons">
                        <button id="sidebar-toggle" class="icon-button" type="button" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </button>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">A collection of resources for data scientists (not only) in bioinformatics</h1>

                    <div class="right-buttons">
                        <a href="print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>
                        <a href="https://github.com/koesterlab/data-science-for-bioinfo" title="Git repository" aria-label="Git repository">
                            <i id="git-repository-button" class="fa fa-github"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="introduction"><a class="header" href="#introduction">Introduction</a></h1>
<p>In this <a href="https://rust-lang.github.io/mdBook/">mdbook</a>, we will collect resources and tips for data scientists in bioinformatics.
It is meant as a quick reference for anybody, from someone getting started without any prior programming knowledge to experienced data scientists.
We initiated its development for a community of bioinformaticians at the DKTK site Essen Düsseldorf, but want to make this open for anybody from the wider local and global data science communities.</p>
<h2 id="contributions"><a class="header" href="#contributions">Contributions</a></h2>
<p>Contributions are very welcome, and can be anything from asking even the smallest question by <a href="https://github.com/koesterlab/data-science-for-bioinfo/issues/new/choose">opening an issue</a>, up to directly suggesting changes or additions via <a href="https://github.com/koesterlab/data-science-for-bioinfo/pulls">pull requests</a>.
The only thing to keep in mind is to keep this collection a welcoming place by following our <a href="https://github.com/koesterlab/data-science-for-bioinfo/blob/main/CODE_OF_CONDUCT.md">code of conduct</a>.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="bioinformatics"><a class="header" href="#bioinformatics">Bioinformatics</a></h1>
<p>For a tool to convert between lots of bioinformatics file formats, see the <a href="bioinformatics/../data_formats/README.html#bioconvert-converting-between-bioinformatics-file-formats"><code>bioconvert</code> description in the <code>data formats</code> section</a>.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="bioconda"><a class="header" href="#bioconda">bioconda</a></h1>
<blockquote>
<p>Bioconda lets you install thousands of software packages related to biomedical research using the conda package manager.</p>
</blockquote>
<p>It has great usage documentation, so check out <a href="https://bioconda.github.io/#usage">how to get started using bioconda on its own website</a>.</p>
<p>The first step (also described and linked to on the website) is to install <a href="bioinformatics/../command_line/conda.html">conda</a>, the package manager you need to use the bioconda channel.
Here, we recommend using the <a href="https://github.com/conda-forge/miniforge#install">mambaforge installation</a>.</p>
<h2 id="conda-basics"><a class="header" href="#conda-basics">conda basics</a></h2>
<p>There is a <a href="bioinformatics/../command_line/conda.html">separate page with some basic <code>conda</code> commands` to get you started</a>.</p>
<h2 id="contributing-to-bioconda"><a class="header" href="#contributing-to-bioconda">contributing to bioconda</a></h2>
<p>While you will nowadays find most bioinformatics packages on bioconda (and checking out a package on bioconda is a great way of checking how often it is downloaded), you will sometimes encounter packages that are not available there.
If you feel comfortable with installing such a tool manually, please consider contributing your solution to the channel in the form of a recipe.
There is a whole <a href="https://bioconda.github.io/contributor/index.html">documentation section dedicated to helping you contribute package recipes to bioconda</a>.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/bioinformatics/bioconda.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="bioconductor"><a class="header" href="#bioconductor">bioconductor</a></h1>
<p>The <a href="https://www.bioconductor.org/">bioconductor website</a> says it best:</p>
<blockquote>
<p>The mission of the Bioconductor project is to develop, support, and disseminate free open source software that facilitates rigorous and reproducible analysis of data from current and emerging biological assays. We are dedicated to building a diverse, collaborative, and welcoming community of developers and data scientists.</p>
<p>Bioconductor uses the R statistical programming language, and is open source and open development.</p>
</blockquote>
<p>Thus, bioconductor provides standardized and reviewed bioinformatics packages, usually with good and extensive documentation.
All of its packages can for example be installed using <code>conda</code> via the <a href="bioinformatics/bioconda.html"><code>bioconda</code> channel</a>.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/bioinformatics/bioconductor.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="coding-tools"><a class="header" href="#coding-tools">Coding tools</a></h1>
<div style="break-before: page; page-break-before: always;"></div><h1 id="useful-coding-conventions"><a class="header" href="#useful-coding-conventions">useful coding conventions</a></h1>
<h2 id="naming-things-the-case-for-snake-case"><a class="header" href="#naming-things-the-case-for-snake-case">naming things: the case for snake case</a></h2>
<p>&quot;Things&quot; can be anything from a file or folder name to the name of a single variable name.
In coding, and in work in general, you have to come up with names for stuff constantly.
The first guide should usually be the style guide of the language you are working in (for example, see the <a href="https://peps.python.org/pep-0008/"><code>python</code> style guide</a>).</p>
<p>For anything where your language's style guide leaves you alone (or that is not in a language with a style guide), we recommend using <code>snake_case</code> or <code>lower_case_with_underscores</code> as a general rule.
Names are very easy to read, because words are clearly separated; and you'll never have to think twice which letter was capitalized.</p>
<p>Also, don't abbreviate unnecessarily.
Typing out a longer variable name is usually something that your integrated development environment (IDE) will do for you (if you're not very quick at it anyway)---what it can't do for you, is remember what the heck that particular variable <code>s</code> stands for...</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/coding_tools/conventions.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="integrated-development-environments-ides"><a class="header" href="#integrated-development-environments-ides">integrated development environments (IDEs)</a></h1>
<p>If all you want to do is done in <code>R</code>, your best choice of IDE is probably <a href="https://posit.co/products/open-source/rstudio/"><code>Rstudio</code></a>.
For all other purposes, you will have to look around and try what works for you.</p>
<h2 id="microsoft-visual-studio-code"><a class="header" href="#microsoft-visual-studio-code">Microsoft Visual Studio Code</a></h2>
<p><a href="https://code.visualstudio.com/">VS code</a> offers a lot of plugins for all common languages, plus lots of useful plugins for other stuff (and all for free).
It has <a href="https://code.visualstudio.com/docs">extensive documentation</a>, but the best way to get started is probably to simply <a href="https://code.visualstudio.com/download">download VS code</a> and jump right in.
Several people in our group currently use it, with the possibility to <a href="https://code.visualstudio.com/docs/remote/ssh">work on a remote project on a server via ssh</a> a very useful and heavily used feature.</p>
<h2 id="jet-brains"><a class="header" href="#jet-brains">Jet Brains</a></h2>
<p>Jet Brains develops a <a href="https://www.jetbrains.com/products/#type=ide-vs">suite of IDEs for different languages</a> (just use the language filter on the right) and offers a <a href="https://www.jetbrains.com/community/education/#students">free version for academic purposes</a>.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/coding_tools/ide.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="licensing"><a class="header" href="#licensing">Licensing</a></h1>
<p>In general, you will have to check with your institution / employer if there are any requirements or limitations for releasing software and code that you develop.
This especially applies to licensing.</p>
<h2 id="open-source-licensing"><a class="header" href="#open-source-licensing">Open Source licensing</a></h2>
<p>To maximize possibilities of code re-use, which should be the default in research, Open Source licenses are the best choice.
For quickly deciding on an Open Source license for your project, use the interactive tool at <a href="https://choosealicense.com/">choosealicense.org</a>.
For a comprehensive list, check out the <a href="https://opensource.org/licenses">Open Source Initiative's community-approved Open Source licenses</a>.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/coding_tools/licensing.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="markdown"><a class="header" href="#markdown">Markdown</a></h1>
<p>Markdown is an plain-text syntax for writing structured texts that are both human-readable as plain-text and machine-readable for automatic parsing into different formats (HTML, PDF, ...).</p>
<p>For a standardized syntax definition, check out the <a href="https://commonmark.org/">CommonMark project</a>. They have a quick <a href="https://commonmark.org/help/">Markdown syntax overview</a> that also links to a 10 minute tutorial.</p>
<p>Markdown is used extensively in many projects, e.g. for writing this <a href="https://rust-lang.github.io/mdBook/format/markdown.html">mdbook</a> or for writing <code>README.md</code> files with <a href="https://github.github.com/gfm/#what-is-github-flavored-markdown-">GitHub flavoured Markdown</a> on GitHub.
As it literally takes no more than 10 minutes to learn the basics and a couple of minutes to look up anything that is a bit more complicated (for example <a href="https://github.github.com/gfm/#tables-extension-">tables in GitHub flavoured Markdown</a>), it is really worthwhile checking it out.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/coding_tools/markdown.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="regular-expressions"><a class="header" href="#regular-expressions">regular expressions</a></h1>
<p>As the <a href="coding_tools/regular_expressions.html#why-regexes-xkcd-reasoning">xkcd comic below</a> suggests, <a href="https://en.wikipedia.org/wiki/Regular_expression">regular expressions</a> are a really powerful tool to systematically search (and replace) strings in very large text files.
If you have ever used search-and-replace in a word processor or a spreadsheet software, you have probably seen cases where just searching for a particular word doesn't do the trick.
This is where regular expressions come in.</p>
<p>To get started, it makes sense to familiarize yourself with the general principal of regexes.
For example, check out the <a href="https://www.regular-expressions.info/quickstart.html">regular-expressions.info Quick Start</a>.
As this introduction also mentions, you will then have to venture on to learn how exactly regular expressions work in the language or tool you want to use.</p>
<h2 id="regular-expression-flavours"><a class="header" href="#regular-expression-flavours">regular expression flavours</a></h2>
<p>Here comes a tables with resources on the exact regular expression syntax in different tools and languages.
These are often referred to as regular expression flavours.</p>
<div class="table-wrapper"><table><thead><tr><th>language / tool</th><th>reference</th><th>further resources</th></tr></thead><tbody>
<tr><td><a href="coding_tools/../command_line/linux_tools.html">linux tools (awk, sed, grep, etc.)</a></td><td><a href="https://www.grymoire.com/Unix/Regular.html">Basic and Extended regexes</a></td><td></td></tr>
<tr><td><a href="coding_tools/../languages/python.html">python</a></td><td><a href="https://docs.python.org/3/library/re.html"><code>re</code> module</a> / <a href="https://docs.python.org/3/howto/regex.html#regex-howto">regex HOWTO</a></td><td><a href="https://pythex.org/">browser-based regex testing</a></td></tr>
<tr><td><a href="coding_tools/../languages/r.html">R</a></td><td><a href="https://stringr.tidyverse.org/articles/regular-expressions.html">regular expression vignette</a></td><td><a href="https://stringr.tidyverse.org/index.html#cheatsheet"><code>stringr</code> package cheat sheet</a></td></tr>
</tbody></table>
</div>
<h2 id="why-regexes-xkcd-reasoning"><a class="header" href="#why-regexes-xkcd-reasoning">why regexes (<a href="https://xkcd.com">xkcd</a> reasoning)</a></h2>
<p><a href="https://xkcd.com/208"><img src="https://imgs.xkcd.com/comics/regular_expressions.png" alt="Wait, forgot to escape a space. Wheeeeee[taptaptap]eeeeee." /></a></p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/coding_tools/regular_expressions.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="user-experience--bug-reports"><a class="header" href="#user-experience--bug-reports">user experience / bug reports</a></h1>
<h2 id="friction-logging"><a class="header" href="#friction-logging">friction logging</a></h2>
<p><a href="https://developerrelations.com/developer-experience/an-introduction-to-friction-logging">Friction logging</a> is a technique to narratively describe you user experience with some process.
This can be the installation and setup of a software, the deployment of a standardized workflow, following along a tutorial, or any other process you go through on your computer.
It can be a very useful document for developers who want to improve their &quot;product&quot;, whatever that is (software, web interface, installation process).
And it is very easy to get started with this, a quick motivation and description for this technique is in this <a href="https://developerrelations.com/developer-experience/an-introduction-to-friction-logging">friction logging blog post by Aja Hammerly</a>.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/coding_tools/user_experience.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="version-control"><a class="header" href="#version-control">version control</a></h1>
<h2 id="git-the-version-control-tool"><a class="header" href="#git-the-version-control-tool">git: the version control tool</a></h2>
<p>For quite some years, <a href="https://git-scm.com/"><code>git</code></a> has become the de-facto standard for distributed version control.
It can be used to granularly track changes to your code base while collaborating in within and across teams.
And its wide adoption means that it integrates well into the graphical user inteface of any modern <a href="coding_tools/ide.html">integrated development environment (IDE)</a>, while you can also use it directly on the <a href="https://git-scm.com/book/en/v2/Getting-Started-The-Command-Line">command line</a>.</p>
<h3 id="learning-git"><a class="header" href="#learning-git">learning git</a></h3>
<ul>
<li><a href="https://happygitwithr.com">happy git with R</a>: A very good and comprehensive, self-paced course for learning how to use git, applying this to coding in R (and integrating it with using the RStudio IDE). </li>
<li><a href="https://git-scm.com/book/en/v2">git book</a>: The <a href="https://git-scm.com/book/en/v2/Getting-Started-About-Version-Control"><code>Getting Started</code></a> and <a href="https://git-scm.com/book/en/v2/Git-Basics-Getting-a-Git-Repository"><code>Git Basics</code></a> chapters should get you going, the following chapters introduce you to more advanced applications.</li>
<li><a href="https://training.github.com/">printout git cheat sheet</a>: A git cheat sheet with the most important git subcommands at one glance. Having this printed out and ready at hand will make looking them up a lot quicker. It is provided by GitHub in a great number of languages.</li>
<li><a href="https://ndpsoftware.com/git-cheatsheet.html">interactive git cheat sheet</a>: For interactively exploring different git subcommands and what they do, with a visualization of different places where code changes can be tracked, this is a very cool tool. It's for playing around or looking stuff up once you feel a bit more comfortable with git.</li>
<li><a href="https://git-scm.com/docs">git command reference</a>: In addition to a full reference of all git subcommands (that you can execute on the command line), this page also provides links to the above mentioned cheat sheets.</li>
</ul>
<h2 id="github-de-facto-standard-online-platform-to-work-with-git-repositories"><a class="header" href="#github-de-facto-standard-online-platform-to-work-with-git-repositories">GitHub: <em>de facto</em> standard online platform to work with git repositories</a></h2>
<p><a href="https://docs.github.com/en">GitHub</a> has become the <em>de facto</em> standard for distributed version control using <code>git</code> repositories.
It has <a href="https://github.com/pricing#compare-features">extensive free features</a>, especially for publicly hosted repositories.
This includes continuous integration resources, even for private repositories.
And GitHub provides extensive documentation.
However, repositories will be hosted on company servers that will not necessarily conform to your country's data regulations.
Thus, for certain projects you might have to use an in-house platform like GitLab (see below).</p>
<h2 id="gitlab-open-source-online-platform-to-work-with-git-repositories"><a class="header" href="#gitlab-open-source-online-platform-to-work-with-git-repositories">GitLab: Open Source online platform to work with git repositories</a></h2>
<p><a href="https://docs.gitlab.com/">GitLab</a> is an <a href="https://opensource.org/osd">Open Source</a> alternative to GitHub, with most of the same features.
As the platform itself is Open Source code, it can be <a href="https://docs.gitlab.com/ee/install/">installed and hosted in-house</a> by an institution.
Thus, it is the go to solution for educational and research instutions that want to avoid having (private) repositories hosted on outside servers.</p>
<h2 id="collaborating-using-git-git-github-and-gitlab-flow"><a class="header" href="#collaborating-using-git-git-github-and-gitlab-flow">collaborating using git: git, GitHub and GitLab flow</a></h2>
<p>For collaborative work on a code base, you need to agree on some basic rules of how to use git and GitHub or GitLab.
Usually, you can pick up on a lot of these conventions by simply using one of the online platforms and by looking at existing projects.
Most of them implicitly use <a href="https://docs.github.com/en/get-started/quickstart/github-flow">GitHub flow</a>, so reading through the explanation linked here will get you a long way.</p>
<p>GitHub flow is a reduction of the more complicated <a href="https://nvie.com/posts/a-successful-git-branching-model/">git flow</a> from 2010 to setup where code is released often and quality is assured by continuous integration.
Should you ever encounter a project where GitHub flow does not provide the level of control you need, <a href="https://docs.gitlab.com/ee/topics/gitlab_flow.html#production-branch-with-gitlab-flow">GitLab flow</a> should get you started.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/coding_tools/version_control.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="command-line"><a class="header" href="#command-line">Command line</a></h1>
<div style="break-before: page; page-break-before: always;"></div><h1 id="bash"><a class="header" href="#bash">bash</a></h1>
<p><code>bash</code> is the most common Linux shell, a command-line environment where you can enter and thus run commands interactively.
You can find extensive resources about it online, and oftentimes searching for your particular problem online will provide suggestions for solutions.</p>
<p>Here are two quick-and-easy introductions that focus on the most important aspects:</p>
<ul>
<li><a href="https://tildesites.bowdoin.edu/~sbarker/unix/">Unix Crash Course by Sean Barker (with xkcd comics;)</a>, and with some infos on server access on the command-line</li>
<li><a href="https://www.csoft.net/docs/course.html">Unix Crash Course by <code>csoft.net</code></a>, with a bit more on file permissions</li>
</ul>
<p>And here are some more detailed resources:</p>
<ul>
<li><a href="https://tldp.org/LDP/Bash-Beginners-Guide/html/index.html">Bash Guide for Beginners</a>: Step-by-step introduction to some of the most important features.</li>
<li><a href="https://tldp.org/LDP/abs/html/">Advanced Bash-Scripting Guide</a>: Extensive reference for (almost) all imaginable things in <code>bash</code>. For example, you can find a <a href="https://tldp.org/LDP/abs/html/refcards.html#AEN22828">table of all string operations in <code>bash</code></a>.</li>
</ul>
<h2 id="bash-scripts"><a class="header" href="#bash-scripts">bash scripts</a></h2>
<p>You can write full scripts in plain text files, with one command per row and including flow controls (like if-else clauses) in <code>bash</code>, so that you can repeat any number of steps easily.
This script can just contain commands, and if it is saved as <code>path/to/some_script.bash</code>, you could execute it with:</p>
<pre><code class="language-bash">bash path/to/some_script.bash
</code></pre>
<p>Instead, you can also specify that it is a <code>bash</code> script within the file, by including the following first line in the file <code>path/to/another_script.bash</code> (if you are interested in what this line does, check out <a href="https://stackoverflow.com/a/21613044">the explanation on stackoverflow</a>):</p>
<pre><code class="language-bash">#!/usr/bin/env bash
</code></pre>
<p>If you then make that file executable for your user with <code>chmod u+x another_script.bash</code>, you can execute it by simply calling the name:</p>
<pre><code class="language-bash">path/to/another_script.bash
</code></pre>
<p>Your shell will know to use the <code>bash</code> from your main environment to execute this script.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/command_line/bash.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="conda--mamba-software-installation"><a class="header" href="#conda--mamba-software-installation">conda / mamba (software installation)</a></h1>
<blockquote>
<p>Conda is an open-source package management system and environment management system that runs on Windows, macOS, and Linux. Conda quickly installs, runs, and updates packages and their dependencies.</p>
</blockquote>
<p>This is how the <a href="https://docs.conda.io/projects/conda/en/latest/index.html"><code>conda</code> documentation</a> puts it.
One more major selling-point: you don't need administrator rights to install conda itself or the to install packages with it.
And for our (bioinformatic) purposes, there are two more very useful aspects:</p>
<ol>
<li>There is <a href="command_line/../bioinformatics/bioconda.html"><code>bioconda</code></a>, a <a href="https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/channels.html"><code>conda</code> channel</a> with pretty much any tool in bioinformatics pre-packaged for you.</li>
<li><code>conda</code> seemlessly <a href="https://snakemake.readthedocs.io/en/latest/snakefiles/deployment.html#integrated-package-management">integrates with <code>snakemake</code></a> for automatic software installation in automated analysis workflows.</li>
</ol>
<h2 id="installing-conda--mamba"><a class="header" href="#installing-conda--mamba">installing conda / mamba</a></h2>
<p>We recommend using the <a href="https://github.com/conda-forge/miniforge#install">mambaforge installation</a>.
<code>mamba</code> is a drop-in replacement command for most of <code>conda</code> (it can do exactly the same things), but it is much faster for solving package dependencies.
This can mean a major speedup for creating and changing environments.
Whenever a command does not actually work with <code>mamba</code>, you can simply use <code>conda</code> instead (it is also installed in the background and is used for any task that's not too slow).</p>
<h2 id="creating-and-using-environments"><a class="header" href="#creating-and-using-environments">creating and using environments</a></h2>
<p><code>conda</code> makes it easy to <a href="https://docs.conda.io/projects/conda/en/latest/user-guide/getting-started.html#managing-environments">create and manage new (named) software environments</a>.
We recommend to create a new (and minimal) software environment whenever possible (for example, in a snakemake workflow, each rule will have its own software environment).
You will then only activate that software environment whenever you need the software installed into it.
By keeping environments minimal, you avoid clashes in the dependencies that software might have (for example, if software X needs python version 3, while software Y still requires python version 2, these cannot be installed into the same environment---but X and Y will happily install in separate environments, which will pull in the required python version).</p>
<p>But, long story short, here's the command to create an environment:</p>
<pre><code class="language-bash">mamba create -n snakemake_env snakemake snakedeploy snakefmt
</code></pre>
<p>You can then activate that environment with:</p>
<pre><code class="language-bash">mamba activate snakemake_env
</code></pre>
<p>And you will be able to call the installed tools, for example <code>snakemake</code>:</p>
<pre><code class="language-bash">snakemake --version
snakemake --help
</code></pre>
<p>If you are done with your current session of using the software, the following gets you back to your original environment (the one you were in before activating <code>snakemake_env</code>):</p>
<pre><code class="language-bash">mamba deactivate
</code></pre>
<p>If you no longer need the entire environment (and the software contained in it) and want to delete it, you can:</p>
<pre><code class="language-bash">mamba env remove -n snakemake_env
</code></pre>
<p>And then, there are <a href="https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html">many other things you can do with environments</a>---just look up the details whenever you need them.</p>
<h2 id="installing-specific-versions-of-software"><a class="header" href="#installing-specific-versions-of-software">installing specific versions of software</a></h2>
<p>Sometimes, you want to install a specific version of a software, for example because something was proven to work with that version in the past.
You can achieve this with:</p>
<pre><code class="language-bash">mamba create -n snakemake_env_7_19_1 snakemake=7.19.1
</code></pre>
<p>And in case you ever need to dive into the details, there is of course <a href="https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/pkg-specs.html#package-match-specifications">documentation on how <code>conda</code> handles version number specifications</a>.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/command_line/conda.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="linux-tools"><a class="header" href="#linux-tools">linux tools</a></h1>
<p>Any linux distribution comes with a set of linux tools for the command line.
In bioinformatics (and data science in general), especially the line by line tools for plain text files are very useful.
Thus, we will introduce a number of these tools here and provide links for useful tutorials and documentation for the more complex ones.
However, you can usually get detailed documentation for any of these tools on the command-line.
Simply write <code>man</code> followed by the tool's name to get its full manual, for example <code>man head</code>.
You can scroll up and down and leave the manual with <code>q</code> to drop back to the command-line (controls are the same as the command-line text file viewer <a href="command_line/linux_tools.html#less-quick-look-at-text-files"><code>less</code></a>).</p>
<h2 id="common-command-line-arguments"><a class="header" href="#common-command-line-arguments">common command-line arguments</a></h2>
<p>Most command-line tools have certain standard command-line arguments, that usually come in a short (one dash <code>-</code> and just one letter) and a long version (two dashes <code>--</code> and a full word):</p>
<ul>
<li><code>-h</code>/<code>--help</code>: display a help message</li>
<li><code>--version</code>: display the version of the tool that you have installed</li>
<li><code>--verbose</code>: provide more verbose output, oftentimes useful for debugging problems</li>
</ul>
<h2 id="head-and-tail"><a class="header" href="#head-and-tail"><code>head</code> and <code>tail</code></a></h2>
<p><code>head -n 5 myfile.txt</code> will display the first five lines of the plain text file <code>myfile.txt</code>.
<code>tail -n 5 myfile.txt</code> will display the last five lines of the plain text file <code>myfile.txt</code>.</p>
<h2 id="less-quick-look-at-text-files"><a class="header" href="#less-quick-look-at-text-files"><code>less</code>: quick look at text files</a></h2>
<p><code>less</code> is a tool for quickly looking at text files on the command line.
Conjure it up with <code>less myfile.txt</code>, scroll up and down with the arrow and page up/down keys, search for something like <code>thisword</code> with <code>/thisword</code> (followed by enter) and quit by hitting <code>q</code>.</p>
<h2 id="grep-line-by-line-plain-text-file-searching"><a class="header" href="#grep-line-by-line-plain-text-file-searching"><code>grep</code>: line by line plain text file searching</a></h2>
<p><code>grep</code> allows you to search for strings and <a href="command_line/../coding_tools/regular_expressions.html">regular expressions</a> in text files.
It will return any line containing a search string.
For example, <code>grep &quot;gene&quot; myfile.txt</code> would return any line containing the string <code>gene</code> in <code>myfile.txt</code>.</p>
<h2 id="sed-line-by-line-plain-text-file-editing"><a class="header" href="#sed-line-by-line-plain-text-file-editing"><code>sed</code>: line by line plain text file editing</a></h2>
<p><code>sed</code> is a great tool to work on plain-text files line by line.
For example, you can easily search and replace patterns using regular expressions.</p>
<ul>
<li><a href="https://www.grymoire.com/Unix/sed.html">sed tutorial for getting started</a></li>
</ul>
<h2 id="awk-line-by-line-editing-of-tabular-plain-text-files"><a class="header" href="#awk-line-by-line-editing-of-tabular-plain-text-files"><code>awk</code>: line by line editing of tabular plain text files</a></h2>
<p><code>awk</code> is a great tool to work on plain-text tabular files (for example tab-separated or comma-separated files, <code>.tsv</code> and <code>.csv</code> files, respectively).
It provides you with direct accessors to individual columns (for example <code>$1</code> for the first column) and with lots of functionality to work with them.</p>
<ul>
<li><a href="https://www.grymoire.com/Unix/Awk.html">awk tutorial for getting started</a></li>
<li><a href="https://www.math.utah.edu/docs/info/gawk_toc.html">comprehensive awk reference</a></li>
</ul>
<h2 id="further-resources"><a class="header" href="#further-resources">further resources</a></h2>
<ul>
<li><a href="https://github.com/stephenturner/oneliners">useful bioinformatics one-liners</a>: these mostly only use basic linux-tools to achieve common bioinformatics tasks</li>
</ul>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/command_line/linux_tools.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="data-analyses"><a class="header" href="#data-analyses">Data analyses</a></h1>
<div style="break-before: page; page-break-before: always;"></div><h1 id="differential-expression-analysis"><a class="header" href="#differential-expression-analysis">differential expression analysis</a></h1>
<p>A classic case of bioinformatic data analysis to check whether expression levels of individual transcripts differs between two (or more) groups of samples.
These can be a treatment group compared to a control group, different time points, or groups differing according any other annotation variable from their metadata.
Up to a certain point, these analyses differ depending on the raw data you start from (for example expression microarrays, full transcriptome RNA-seq data, or specialized RNA sequencing methods).
Thus, we have <a href="data_analyses/differential_expression.html#data-types">one section per data type</a> here, and include linkouts to standardized (snakemake) workflows wherever available, which will make it much quicker for you to get started with your analysis, and much easier to update and/or redo it later.
However, after the initial data processing, differential expression analysis is usually performed by the same (statistical) methods.
Thus, we have <a href="data_analyses/differential_expression.html#differential-expression-analysis-tools">another section dedicated to differential expression analysis tools</a>.
Finally, transcripts (or genes) that have been found to be differentially expressed, can be analyzed further---for example via gene set or enrichment or pathway perturbation analysis.
We list some tools for these tasks in the <a href="data_analyses/differential_expression.html#analysis-of-differentially-expressed-transcripts">section <code>analysis of differentially expressed transcripts</code></a></p>
<h2 id="data-types"><a class="header" href="#data-types">data types</a></h2>
<h3 id="microarray-data"><a class="header" href="#microarray-data">microarray data</a></h3>
<p>While fewer and fewer new microarray data sets are being generated, a lot of existing microarray data sets can be re-analyzed for new research questions.
If you want to get started with such analyses, there is a great hands-on tutorial by Greg Tucker-Kellogg: <a href="https://gtk-teaching.github.io/Microarrays-R/">Introduction to gene expression microarray analysis in R and Bioconductor</a>.
It includes an introduction to Bioconductor, works on publicly available data and is completely self-contained.
You can work through it at your own pace and all it requires is some basic R knowledge---for the latter, see our section on <a href="data_analyses/../languages/r.html#learning-programming-with-r">learning R</a>.</p>
<p>Once you are familiar with the data type and the analysis, there's a <a href="https://snakemake.github.io/snakemake-workflow-catalog/?usage=zifornd/arrays">standardized snakemake workflow for microarray analysis</a> by James Ashmore.
While we haven't used this workflow ourselves (but might in the future, we'll try to update the info here if we do), this workflow is:</p>
<ul>
<li>is standardized, so:
<ul>
<li>the above link should take you through all of the setup steps you will need to run this workflow on your own data</li>
<li>it is well-structured, so modifying it to your needs should be relatively easy</li>
</ul>
</li>
<li>based on a thoroughly peer reviewed walk-through manuscript on microarray analysis: https://doi.org/10.12688/f1000research.8967.2</li>
</ul>
<h3 id="rna-seq-data"><a class="header" href="#rna-seq-data">RNA-seq data</a></h3>
<p>TODO: provide good self-contained tutorial for working with RNA-seq data</p>
<p>There are several standardized snakemake workflows, the following ones are maintained by the Kösterlab and associated people: </p>
<ul>
<li><a href="https://snakemake.github.io/snakemake-workflow-catalog/?usage=snakemake-workflows/rna-seq-kallisto-sleuth">a workflow using <code>kallisto</code> for quantifying transcript expression and <code>sleuth</code> (see below)</a> for determining differentially expressed transcripts</li>
<li><a href="https://snakemake.github.io/snakemake-workflow-catalog/?usage=snakemake-workflows/rna-seq-star-deseq2">a workflow using <code>Star</code> to align reads to the transcriptome and <code>DESeq2</code> (see below)</a> for determining differentially expressed transcripts</li>
</ul>
<h2 id="differential-expression-analysis-tools"><a class="header" href="#differential-expression-analysis-tools">differential expression analysis tools</a></h2>
<h3 id="sleuth"><a class="header" href="#sleuth">sleuth</a></h3>
<p>TODO: add some info
<a href="https://pachterlab.github.io/sleuth/walkthroughs"><code>sleuth</code> walkthroughs for different analyses types</a></p>
<h3 id="deseq2"><a class="header" href="#deseq2">DeSEQ2</a></h3>
<p>TODO: add some info
<a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">extensive HTML documentation page for <code>DESeq2</code></a>
<a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">main bioconductor package page for <code>DESeq2</code></a></p>
<h3 id="edger"><a class="header" href="#edger">edgeR</a></h3>
<p>TODO: add some info
<a href="https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf">extensive PDF user guide for <code>edgeR</code></a>
<a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">main bioconductor package page for <code>edgeR</code></a></p>
<h2 id="analysis-of-differentially-expressed-transcripts"><a class="header" href="#analysis-of-differentially-expressed-transcripts">analysis of differentially expressed transcripts</a></h2>
<h3 id="gene-set-enrichment"><a class="header" href="#gene-set-enrichment">gene set enrichment</a></h3>
<h3 id="pathway-enrichment"><a class="header" href="#pathway-enrichment">pathway enrichment</a></h3>
<h3 id="pathway-perturbation"><a class="header" href="#pathway-perturbation">pathway perturbation</a></h3>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/data_analyses/differential_expression.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><p>In this section, tools for downloading data that may include raw sequencing data, genes, genomes, annotations, etc. are gathered.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="ncbi-datasets-gathering-data-across-ncbi-databases"><a class="header" href="#ncbi-datasets-gathering-data-across-ncbi-databases">ncbi-datasets: gathering data across NCBI databases</a></h1>
<p><a href="https://github.com/ncbi/datasets">ncbi-datasets</a> is a very-well documented open source tool from NCBI (National Center for Biotechnology Information) for gathering various kinds of data including genes, genomes, annotations etc. belonging to the desired species.
As a side note, it has specific options for retrieving SARS-CoV-2 data.</p>
<h2 id="installation"><a class="header" href="#installation">Installation</a></h2>
<p>It is offered as a conda package and can be installed via:</p>
<pre><code>conda create -n ncbi-download -c conda-forge ncbi-datasets-cli
</code></pre>
<h2 id="example-usage"><a class="header" href="#example-usage">Example usage</a></h2>
<p>Retrieval of human reference genome:</p>
<pre><code>datasets download genome taxon human --reference --filename human-reference.zip
</code></pre>
<p>Retrieval of genomic assemblies for a specific SARS-CoV-2 lineage with the host being homo sapiens:</p>
<pre><code>datasets download virus genome taxon SARS-CoV-2 --complete-only --filename B.1.525.zip --lineage B.1.525 --host &quot;homo sapiens&quot;
</code></pre>
<p>A very nice illustration that shows the overall structure of data download via ncbi-datasets:
<img src="https://camo.githubusercontent.com/126831e1b7a97c746e0007e6f974f470d68f56d2aec8f31986e9158d1ca7fb73/68747470733a2f2f7777772e6e6362692e6e6c6d2e6e69682e676f762f64617461736574732f646f63732f76322f64617461736574735f736368656d615f636f6d706c6574655f7631342e706e67" alt="ncbi-datasets-structure" /></p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/data_downloads/ncbi_datasets_cli.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="data-formats-and-tools"><a class="header" href="#data-formats-and-tools">Data formats and tools</a></h1>
<p>This section gathers common bioinformatics data formats, where you can find specifications of them and which tools you can use to handle them sanely.</p>
<h2 id="bioconvert-converting-between-bioinformatics-file-formats"><a class="header" href="#bioconvert-converting-between-bioinformatics-file-formats">bioconvert: converting between bioinformatics file formats</a></h2>
<p>There's a plethora of bioinformatics file formats, and you often need to convert from on to another.
<a href="https://bioconvert.readthedocs.io/en/dev/"><code>bioconvert</code></a> is a tool that integrates as many of these conversions as possible.
It's a community effort, so feel free <a href="https://github.com/bioconvert/bioconvert/issues/1">to contribute a converter</a> if the one you need doesn't exist, yet, but you found (or implemented) a tool that can do it.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="sequencing-data-fastq"><a class="header" href="#sequencing-data-fastq">sequencing data (fastq)</a></h1>
<h2 id="tools-for-downloading-from-repositories"><a class="header" href="#tools-for-downloading-from-repositories">tools for downloading from repositories</a></h2>
<h3 id="fastq-dl"><a class="header" href="#fastq-dl">fastq-dl</a></h3>
<p><a href="https://github.com/rpetit3/fastq-dl">fastq-dl</a> lets you download FASTQ data from the European Nucleotide Archive or the Sequence Read Archive repositories.
It allows you to use any ENA/SRA accession number, for a <code>Study</code>, a <code>Sample</code>, an <code>Experiment</code>, or a <code>Run</code>.</p>
<h2 id="splitting-fastq-data"><a class="header" href="#splitting-fastq-data">Splitting FASTQ data</a></h2>
<h3 id="fastqsplitter"><a class="header" href="#fastqsplitter">fastqsplitter</a></h3>
<p><a href="https://github.com/LUMC/fastqsplitter">Fastqsplitter</a> is a tool that allows you to split FASTQ files into a desired number of output files <em>evenly</em>.</p>
<p>Installation of fastqsplitter via conda:</p>
<pre><code>mamba create -n fastqsplitter -c bioconda fastqsplitter
</code></pre>
<p>Usage of fastqsplitter:</p>
<pre><code>fastqsplitter -i input.fastq.gz -o output_1.fastq.gz -o output_2.fastq.gz -o output_3.fastq.gz
</code></pre>
<p>In the example above, fastqsplitter divides the input fastq into 3 evenly distributed fastq files through inferring this number from the outputs that are specified by the user.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/data_formats/fastq.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="comma-separated-or-tab-separated-value-files-csvtsv"><a class="header" href="#comma-separated-or-tab-separated-value-files-csvtsv">comma-separated or tab-separated value files (CSV/TSV)</a></h1>
<h2 id="tools"><a class="header" href="#tools">tools</a></h2>
<p>These can also often be handled with <a href="data_formats/../command_line/linux_tools.html">standard linux command line tools</a>.
However, these usually do not respect header lines and require complicated extra code to deal with those.
The following tools all respect header lines.
However, always be aware of the trade-off between having lots of functionality (or specific functionality like a <code>join</code> command) and a cost in processing speed or memory usage that this might incur.</p>
<h3 id="miller"><a class="header" href="#miller">miller</a></h3>
<blockquote>
<p><a href="https://miller.readthedocs.io">Miller</a> is a command-line tool for querying, shaping, and reformatting data files in various formats including CSV, TSV, JSON, and JSON Lines.</p>
</blockquote>
<p>Or, to rephrase this: <a href="https://miller.readthedocs.io">miller</a> is the swiss army knife of CSV/TSV files, you should be able to do almost anything with it.
It uses the linux pipe philosophy, <a href="https://miller.readthedocs.io/en/latest/streaming-and-memory/">streaming through input and output</a> wherever possible, to also reduce memory usage.
And it employs <a href="https://miller.readthedocs.io/en/latest/cpu/">multithreading to a certain degree</a>, especially for (de-)compression and for multiple commands in its <code>then</code>-chains.</p>
<h3 id="xsv"><a class="header" href="#xsv">xsv</a></h3>
<p><a href="https://github.com/burntsushi/xsv">xsv</a> is not as versatile as miller, but blazingly fast.
It also follows the linux pipe philosophy and streams wherever possible.
If it can do what you are looking for, it is probably the fastest option out there.</p>
<h3 id="csvkit"><a class="header" href="#csvkit">csvkit</a></h3>
<p><a href="https://csvkit.readthedocs.io/">csvkit</a> can for example join multiple csv files by a common column at once, using <code>csvjoin</code>, whereas other tools will require pipe chaining of multiple calls to their join command for this.
However, it is <a href="https://faraday.ai/blog/how-we-made-our-csv-processing-142x-faster/">a lot slower than <code>xsv</code></a>.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/data_formats/tsv.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="yaml-syntax"><a class="header" href="#yaml-syntax">YAML syntax</a></h1>
<blockquote>
<p>YAML is a human friendly language for the communication of data between people and computers.
The language is rich enough to represent almost any conceivable data structure or object inside a running computer program, as a readable plain text file. This type of language is often called a “serialization language” and YAML is definitely one of those.
But unlike most serialization languages, the YAML format tries hard to make simple things really simple.</p>
</blockquote>
<p>This is how the <a href="https://yaml.com/">YAML data project</a> puts it, and it's spot on.
Thus, for anything where we need configuration files that should be readable by both humans and machines, we use the YAML language.</p>
<h2 id="yaml-resources"><a class="header" href="#yaml-resources">YAML resources</a></h2>
<ul>
<li><a href="https://www.yaml.info/learn/index.html">15 minute YAML tutorial for the most important aspects</a></li>
<li><a href="https://www.yaml.info/learn/bestpractices.html">best practices</a></li>
<li><a href="https://yaml.org/spec/1.2/">ALL the gory details in the official specification</a></li>
</ul>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/data_formats/yaml.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="data-visualization"><a class="header" href="#data-visualization">Data visualization</a></h1>
<p>If you are interested in a conceptual framework for data visualizations, read up on the <a href="data_presentation/grammar_of_graphics.html">grammar of graphics</a>.
Otherwise, jump right into one of the topics covered here.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="color-palettes"><a class="header" href="#color-palettes">color palettes</a></h1>
<p>Color palettes are a central tool in data visualization.
Thus, it's important to learn <a href="https://blog.datawrapper.de/which-color-scale-to-use-in-data-vis/">which color scale to use when</a>.
And for any visualization using color, please make sure to use colorblind safe palettes:</p>
<h2 id="color-blind-friendly-palettes"><a class="header" href="#color-blind-friendly-palettes">color blind friendly palettes</a></h2>
<p>If you just want a color blind color scale, here are some great resources:</p>
<ul>
<li><a href="https://jfly.uni-koeln.de/color/#pallet">Okabe Ito colorblind safe palette</a>:
In addition to the linked palette, the website also gives good recommendations for colorblind proofing figures.</li>
<li><a href="https://colorbrewer2.org">Color Brewer 2</a>:
With this interactive color palette chooser, you can select sequential, diverging and qualitative colo scales.
It has a checkbox to restrict results to colorblind safe palettes.
In R, these color palettes are directly available in the <a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html"><code>ggplot2</code> package via the <code>scale_colour_brewer()</code></a>, or generally in R via the <a href="https://cran.r-project.org/web/packages/RColorBrewer/index.html"><code>RColorBrewer</code> package</a>.</li>
</ul>
<p>If you're looking for a thorough introduction to the topic, datawrapper has you covered with a three-part blog series:</p>
<ol>
<li><a href="https://blog.datawrapper.de/colorblindness-part1/">How your colorblind and colorweak readers see your colors</a></li>
<li><a href="https://blog.datawrapper.de/colorblindness-part2/">What to consider when visualizing data for colorblind readers</a></li>
<li><a href="https://blog.datawrapper.de/colorblindness-part3/">What’s it like to be colorblind</a></li>
</ol>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/data_presentation/color_palettes.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="grammar-of-graphics"><a class="header" href="#grammar-of-graphics">grammar of graphics</a></h1>
<blockquote>
<p>Basically, a grammar of graphics is a framework which follows a layered approach to describe and construct visualizations or graphics in a structured manner. </p>
</blockquote>
<p>This is how Dipanjan Sarkar puts it in his <a href="https://towardsdatascience.com/a-comprehensive-guide-to-the-grammar-of-graphics-for-effective-visualization-of-multi-dimensional-1f92b4ed4149">introduction to the grammar of graphics on towardsdatascience</a>.
While the most popular implementation of this concept is in the <a href="https://ggplot2.tidyverse.org/"><code>ggplot2</code> package</a> (part of the <code>tidyverse</code> in <code>R</code>), implementations in other languages have followed.
For example, the above linked blog post uses <a href="https://plotnine.readthedocs.io/en/stable/">the python implementation in the package <code>plotnine</code></a>.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/data_presentation/grammar_of_graphics.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="streamlitio"><a class="header" href="#streamlitio">streamlit.io</a></h1>
<p>For building apps and demos to showcase your project and/or project results, this Python library is very easy to implement an app in a straighforward way: https://streamlit.io/ </p>
<p>Using Streamlit:</p>
<ul>
<li>Project reports can be presented with sidebars, widgets..etc. and plots can be inserted (support for Vega-lite, Graphviz and many others) </li>
<li>Algorithm development could be showcased to nonbioinformaticians with some mockup information</li>
<li>A tutorial to create your first app: https://docs.streamlit.io/library/get-started/create-an-app</li>
</ul>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/coding_tools/interactive_reporting.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="languages"><a class="header" href="#languages">Languages</a></h1>
<p>In this section, we'll gather useful resources for different programming languages.
Each subsection starts out with learning resources to get you started.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="python"><a class="header" href="#python">python</a></h1>
<blockquote>
<p><a href="https://www.python.org/">Python</a> is a programming language that lets you work quickly and integrate systems more effectively.</p>
</blockquote>
<p>That is how <a href="https://www.python.org/"><code>python.org</code></a> puts it.
We can maybe add that python comes with a great Open Source community around it, including a large package ecosystem with packages for everything from data science, to bioinformatics or machine learning.
We suggest some below, but whenever you need to do something in python (like parsing a particular file format or running some machine learning method), it usually makes sense to go first check whether there is a package that will provide you with all the functionality for this.</p>
<h2 id="learning-programming-with-python"><a class="header" href="#learning-programming-with-python">learning programming with python</a></h2>
<p>The <a href="https://software-carpentry.org/">software carpentries</a> provide a great <a href="https://swcarpentry.github.io/python-novice-inflammation/"><strong>Programming with Python</strong> course</a> that you can also work through self-paced.
It includes linkouts for how to set up for the course, so you should be able to get yourself up and learning in no time.</p>
<h2 id="python-coding-style"><a class="header" href="#python-coding-style">python coding style</a></h2>
<p><code>python</code> is known for its <a href="https://peps.python.org/pep-0008/">rigorous style guide</a>, some of which (like indentation) is also rigorously enforced.
There's a lot to learn from the style guide, so it's well worth a read!</p>
<h2 id="pandas-the-standard-for-tabular-data"><a class="header" href="#pandas-the-standard-for-tabular-data">pandas: the standard for tabular data</a></h2>
<p><a href="https://pandas.pydata.org/"><code>pandas</code></a> is the current <em>de facto</em> standard package for working with tabular data (a.k.a. 95% of all of data science / bioinformatics).
Its documentation provides a number of <a href="https://pandas.pydata.org/docs/getting_started/index.html">resources for <strong>Getting Started</strong></a>, just see which material fits you best (for example, have you used <code>R</code> or <code>Excel</code> in the past...?).</p>
<p>Generally, there are often multiple ways of achieving the same thing with <code>pandas</code>.
To be consistent and make your life a bit easier, there's a <a href="https://github.com/tdpetrou/Minimally-Sufficient-Pandas">suggestion for <strong>Minimally Sufficient Pandas</strong></a>, a subset of all pandas functionality.
It makes sense to familiarize with that subset early on while learning to use <code>pandas</code>.</p>
<h2 id="polars-the-future-for-tabular-data"><a class="header" href="#polars-the-future-for-tabular-data">polars: the future for tabular data</a></h2>
<p>There's also a newer <code>python</code> library for tabular data, called <a href="https://pola-rs.github.io/polars-book/user-guide/"><code>polars</code></a>.
It has a more efficient implementation and a cleaner usage syntax.
However, it is not as widely adopted and will not work on some older computers, so the current go-to recommendation is still <code>pandas</code>.
And while <code>polars</code> is probably the future for tabular data in python, a transition from <code>pandas</code> to <code>polars</code> won't be too difficult.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/languages/python.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="r"><a class="header" href="#r">R</a></h1>
<h2 id="learning-programming-with-r"><a class="header" href="#learning-programming-with-r">learning programming with R</a></h2>
<p>If you are new to programming and want to start learning using R, the free online book <a href="https://rstudio-education.github.io/hopr/">Hands-On Programming with R (HOPR)</a> is a great place to start.
It really starts from scratch, including information on <a href="https://rstudio-education.github.io/hopr/starting.html">how to install R and Rstudio</a> and getting you acquainted with basic principles of programming through hands-on excercises.</p>
<h2 id="quick-introduction-to-r"><a class="header" href="#quick-introduction-to-r">quick introduction to R</a></h2>
<p>If you have programmed or written scripts before, this <a href="https://gtk-teaching.github.io/Intro-to-R/">quick <code>Intro to R</code></a> by Greg Tucker-Kellogg should get you up to speed in a few hours.
It is a self-contained course, that you can easily do at your own pace.
And it introduces all the necessary basics for starting to work with R.
For installation, I nevertheless recommend the information on <a href="https://rstudio-education.github.io/hopr/starting.html">how to install R and Rstudio from HOPR</a>.</p>
<h2 id="stat545-data-wrangling-exploration-and-analysis-with-r"><a class="header" href="#stat545-data-wrangling-exploration-and-analysis-with-r">Stat545: Data wrangling, exploration, and analysis with R</a></h2>
<p>If you are looking for a very detailed and granular introduction to data analysis with R, <a href="https://stat545.com">STAT545: Data wrangling, exploration, and analysis with R</a> by Jenny Brian <em>et al.</em> provides a thorough and accessible course.
You can work through it self-paced.</p>
<h2 id="interactive-learning-via-swirl"><a class="header" href="#interactive-learning-via-swirl">interactive learning via <code>swirl</code></a></h2>
<p>Interactive learning within R command line. From beginner to advanced. User friendly.
<a href="https://swirlstats.com/">swirl</a></p>
<h2 id="r4ds-r-for-data-science"><a class="header" href="#r4ds-r-for-data-science">R4DS: R for Data Science</a></h2>
<p>If you have programmed before and want to learn how to use R for data analysis, the free online book <a href="https://r4ds.hadley.nz/">R for Data Science</a> is all you need.
It introduces both concepts and tools in a very accessible way, including example code and coding excercises.
It is based on the <a href="languages/r.html#tidyverse">tidyverse</a> packages.</p>
<h2 id="tidyverse"><a class="header" href="#tidyverse">tidyverse</a></h2>
<p>Use the <a href="https://www.tidyverse.org/">tidyverse</a> for modern, vectorized, and efficient code, with excellent documentation (and cool package badges;).
If you want a thorough introduction, the <a href="languages/r.html#r4ds-r-for-data-science">R4DS book</a> is there for you.
And if you want to jump right in, here's a table of the most important packages and their online documentation, in the order that you will probably most often use them in a project:</p>
<h3 id="important-packages"><a class="header" href="#important-packages">important packages</a></h3>
<div class="table-wrapper"><table><thead><tr><th>name</th><th>online docs (cheatsheets!)</th><th>purpose</th></tr></thead><tbody>
<tr><td>readr</td><td><a href="https://readr.tidyverse.org/">https://readr.tidyverse.org/</a></td><td>read in (rectangular) data (csv, tsv, xslx, ...)</td></tr>
<tr><td>tidyr</td><td><a href="https://tidyr.tidyverse.org/">https://tidyr.tidyverse.org/</a></td><td>get data into tidy format</td></tr>
<tr><td>stringr</td><td><a href="https://stringr.tidyverse.org/">https://stringr.tidyverse.org/</a></td><td>string manipulation</td></tr>
<tr><td>dplyr</td><td><a href="https://dplyr.tidyverse.org/">https://dplyr.tidyverse.org/</a></td><td>transform data (in tidy format) with sql-like syntax</td></tr>
<tr><td>ggplot2</td><td><a href="https://ggplot2.tidyverse.org/">https://ggplot2.tidyverse.org/</a></td><td>easy and versatile plotting with the grammar of graphics</td></tr>
</tbody></table>
</div>
<h3 id="jump-right-in"><a class="header" href="#jump-right-in">jump right in</a></h3>
<p>If you do want to jump right in, depending on you previous exposure to the tidyverse, I recommend to at least:</p>
<ol>
<li>Get all the cheatsheets and have those ready on your machine or even printed out.
Either from the online documentation linked above, or from the Rstudio compilation of all tidyverse cheatsheets.</li>
<li>Read through the introductions of the online vignettes above, and the conceptual vignettes that they link to.
They do a very good job at explaining basic concepts that will make writing analyses with the tidyverse a lot easier and more intuitive.</li>
</ol>
<p>The most important concepts are probably:</p>
<ol>
<li>Always get your data into a <strong>tidy format</strong> first.
How tidy data should look and why, is described <a href="https://tidyr.tidyverse.org/articles/tidy-data.html">in this tidy data vignette</a>.</li>
<li>When working in the tidyverse, you will always use <strong>pipes</strong>.
There is a <a href="https://r4ds.had.co.nz/pipes.html">good introductory chapter on pipes in the R4DS online book</a>.</li>
<li>There’s also a <a href="https://dplyr.tidyverse.org/articles/two-table.html">vignette on how to work with two tables</a>.
For example, if you want to filter one table to only include entries also present in a second table.
Or if you want to use a second table, to add annotation fields to a first table.</li>
</ol>
<h3 id="style-guide"><a class="header" href="#style-guide">style guide</a></h3>
<p><a href="https://style.tidyverse.org/">https://style.tidyverse.org</a></p>
<p>A detailed and well-written style guide for using R with the tidyverse.
It covers topics like good names, syntax recommendations and best practices for error messages.
Many of the things covered here, can also be applied in other programming languages.</p>
<h3 id="styler-automatic-formatting"><a class="header" href="#styler-automatic-formatting">styler: automatic formatting</a></h3>
<p><a href="https://styler.r-lib.org/">https://styler.r-lib.org</a></p>
<p>Automatic formatting according to tidyverse style guide.
This is also available for <a href="languages/r.html#alternative-ide-r-reditorsupport-extension-for-vs-code">VS Code via the <code>R</code> (or <code>REditorSupport</code>) extensions</a>.</p>
<h2 id="rstudio-an-integrated-development-environment-tailor-made-for-r"><a class="header" href="#rstudio-an-integrated-development-environment-tailor-made-for-r">Rstudio: an integrated development environment tailor-made for R</a></h2>
<p>Rstudio is an IDE dedicated to R, and probably the most comprehensive IDE for this language and very easy to get going with a productive workflow.
You can get the free desktop version for your operating system on posit's <a href="https://posit.co/download/rstudio-desktop/#download">Rstudio download page</a>.</p>
<p>In Rstudio, you have a console with an interactive R session on the bottom left (like running R on the command line).
Here, you can always conjure up the help information of any function by putting a question mark in front of it (for example <code>?mutate()</code>) and hitting enter.
This will open the respective vignette in the bottom right window.
There, you will first find a description of the function and each of its arguments, and further down some quick examples on tibbles (data frames) that are immediately useable in the interactive session.
So just copy and paste them, and start modifying them to your needs.</p>
<p>If you are using Ubuntu as your operating system and want to install R (and R packages) through conda, here's a useful trick:</p>
<ul>
<li>Getting Rstudio to use a conda environment with your R installation on Ubuntu: <a href="https://gist.github.com/dlaehnemann/8126c903d560307cb246fa26b52d10de">https://gist.github.com/dlaehnemann/8126c903d560307cb246fa26b52d10de</a></li>
</ul>
<h3 id="alternative-ide-r-reditorsupport-extension-for-vs-code"><a class="header" href="#alternative-ide-r-reditorsupport-extension-for-vs-code">alternative IDE: <code>R</code> (<code>REditorSupport</code>) extension for VS Code</a></h3>
<p><a href="https://github.com/REditorSupport/vscode-R">https://github.com/REditorSupport/vscode-R</a></p>
<p>Provides full language support for R in VS Code, including code completion, formatting, syntax highlighting, plot and data views, etc. This should make VS Code configurable in the same way that RStudio works.</p>
<h2 id="error-handling"><a class="header" href="#error-handling">error handling</a></h2>
<p>A <a href="https://style.tidyverse.org/error-messages.html#error-messages">recommendation in the <code>tidyverse</code> style guide</a> points to <a href="https://cli.r-lib.org/index.html"><code>cli</code> package</a> that provides <a href="https://cli.r-lib.org/reference/cli_abort.html">nice functionality for useful error messages via <code>cli_abort</code></a>.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/languages/r.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="rust"><a class="header" href="#rust">Rust</a></h1>
<p><a href="https://www.rust-lang.org/">Rust</a> is a good language to write (bioinformatics) software in, with a focus on efficiency and reliability.
You will be able to analyze large (bioinformatics) data sets very quickly, with inherently safe memory management and easy parallelization.
It has <a href="https://www.rust-lang.org/learn">extensive documentation</a> and a very active <a href="https://www.rust-lang.org/community">user community</a>.</p>
<h2 id="learning-rust"><a class="header" href="#learning-rust">learning Rust</a></h2>
<p>The language website already has a lot of resources, for example:</p>
<ul>
<li><a href="https://www.rust-lang.org/learn/get-started">Getting Started with Rust</a></li>
<li><a href="https://www.rust-lang.org/learn">Learn Rust</a>, with all of the &quot;official&quot; learning materials</li>
<li>the <a href="https://play.rust-lang.org/">Rust Playground</a>, for trying out code in an interactive browser session</li>
</ul>
<p>And then, there is a public 4-day course for learning Rust, called <a href="https://google.github.io/comprehensive-rust/">&quot;Comprehensive Rust&quot;</a>.
While this does require that you already have some programming experience and is originally meant as a course to be taught, you can nevertheless work through it at your own pace.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/languages/rust.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="statistics"><a class="header" href="#statistics">Statistics</a></h1>
<p>For a general understanding of <strong>Probability, Mathematical Statistics, Stochastic Processes</strong>, I often turn to <a href="http://www.randomservices.org/random/index.html"><strong>Random</strong> on randomservices.org</a>, a project by Kyle Siegrist.
It has great explanations and little apps to play around with certain processes and probability distributions.
Especially the <a href="http://www.randomservices.org/random/dist/index.html">explanations of probability distributions</a> are very useful and often much better than what you can find on Wikipedia.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="bayesian-statistics"><a class="header" href="#bayesian-statistics">bayesian statistics</a></h1>
<p>You want a basic understanding or intuition of bayesian statistics (as opposed to frequentist statistics)?
The presentation <a href="https://faculty.washington.edu/kenrice/BayesIntroClassEpi2018.pdf"><strong>Bayesian Statistics (a very brief introduction)</strong></a> by Ken Rice could be a good starting point.</p>
<h2 id="learning-bayesian-statistics-with-r"><a class="header" href="#learning-bayesian-statistics-with-r">learning bayesian statistics with R</a></h2>
<p>There is <a href="https://www.bayesrulesbook.com/"><strong>Bayes Rules! An Introduction to Applied Bayesian Modeling</strong></a> by Alicia A. Johnson, Miles Q. Ott, and Mine Dogucu.
They provide a thorough introduction to bayesian statistics, all with excercises in <code>R</code> (even including their own dedicated <code>R</code> package <code>bayesrule</code> to accompany the book).</p>
<p>A slightly shorter introduction to bayesian statistics in R can be found in the <a href="https://learningstatisticswithr.com/book/bayes.html"><strong>Bayesian statistics</strong> chapter</a> of the book <a href="https://learningstatisticswithr.com/book/">Learning Statistics with R</a> by Danielle Navarro.
It is aimed at a psychology audience, but the examples should all be generic enough to follow.</p>
<p>Or you could check out this <a href="https://www.r-bloggers.com/2019/05/bayesian-models-in-r-2/"><strong>Bayesian models in R</strong></a> <code>r-bloggers</code> post by Francisco Lima.
It also has R code to follow along.</p>
<h2 id="learning-bayesian-statistics-with-python"><a class="header" href="#learning-bayesian-statistics-with-python">learning bayesian statistics with python</a></h2>
<p>You would rather learn bayesian statistics using python?
Look no further, here's <a href="https://allendowney.github.io/ThinkBayes2/"><strong>Think Bayes 2</strong></a> by Allen B. Downey.
It comes with a Jupyter Notebook per chapter, so you can directly run (and play with) the code in the book.</p>
<h2 id="causal-inference-and-bayesian-statistics"><a class="header" href="#causal-inference-and-bayesian-statistics">Causal inference and Bayesian statistics</a></h2>
<p><a href="https://xcelab.net/rm/">Richard McElreath</a> has a nice introduction to causal inference and applying it to Bayesian methods called <a href="https://www.routledge.com/Statistical-Rethinking-A-Bayesian-Course-with-Examples-in-R-and-STAN/McElreath/p/book/9780367139919">Statistical Rethinking</a>. It is also available as a set of <a href="https://www.youtube.com/playlist?list=PLDcUM9US4XdPz-KxHM4XHt7uUVGWWVSus">lectures on youtube</a>, updated annually. There is also an R package <code>rethinking</code> to go with the book. </p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/statistics/bayesian.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="frequentist-statistics"><a class="header" href="#frequentist-statistics">frequentist statistics</a></h1>
<h2 id="learning-statistical-modeling"><a class="header" href="#learning-statistical-modeling">learning statistical modeling</a></h2>
<p>Susan Holmes and Wolfgang Huber have written the very comprehensive book <a href="https://www.huber.embl.de/msmb/"><strong>Modern Statistics for Modern Biology</strong></a>, and you can read it all online for free.
Most importantly though, it has example R code and excercises to follow along.
Thus, if you want to learn all about frequentist statistics---with application in modern biology---this is the perfect resource.</p>
<h2 id="all-the-hypothesis-tests"><a class="header" href="#all-the-hypothesis-tests">all the hypothesis tests</a></h2>
<p>The <a href="http://www.biostathandbook.com/"><strong>Handbook of Biological Statistics</strong></a> by John H. McDonald has great explanations of all the statistical tests you could ever need in biomedical research.
And it has this <a href="http://www.biostathandbook.com/testchoice.html">amazing tabular overview of all of the statistical tests</a>, so you can easily choose the one you need for your data set at hand.</p>
<p>If you want to understand a bit more about how the theory behind a number of these tests is very similar (and thus not all too hard to actually understand), you can check out <a href="https://lindeloev.github.io/tests-as-linear/"><strong>Common statistical tests are linear models</strong></a> by Jonas Kristoffer Lindeløv.
The article has a great summary table and provides example code in R.
So it's also a great resource if you just want quick example code for running a particular statistical test in R.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/statistics/frequentist.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="workflow-management"><a class="header" href="#workflow-management">Workflow management</a></h1>
<div style="break-before: page; page-break-before: always;"></div><h1 id="snakemake"><a class="header" href="#snakemake">Snakemake</a></h1>
<p><a href="https://snakemake.readthedocs.io">Snakemake</a> is a workflow management system to create reproducible and scalable data analyses in a human readable, Python based language. 
You define individual analysis steps as <code>rules</code> with a defined <code>input</code> and <code>output</code>, where dependencies are defined implicitly when the <code>input</code> of one rule matches the <code>output</code> of another rule.</p>
<p>If you want to get to know its core features, there is a <a href="https://snakemake.readthedocs.io/en/latest/tutorial/tutorial.html">great snakemake tutorial</a> that is easy to set up and has excersises to work through at your own pace.
And once you have a general understanding of how snakemake works, you could for example try to use one of the standardized <a href="workflows/snakemake.html#snakemake-workflows-reusable-analysis-workflows">snakemake workflows</a>.</p>
<h2 id="snakemake-workflows-reusable-analysis-workflows"><a class="header" href="#snakemake-workflows-reusable-analysis-workflows">Snakemake workflows: reusable analysis workflows</a></h2>
<p>The <a href="https://snakemake.readthedocs.io/en/latest/snakefiles/modularization.html#modules">snakemake module system</a> allows for flexibly combining different existing workflows for more complex analyses.
It is also used for the deployment of standardized workflows that you can find in the <a href="https://snakemake.github.io/snakemake-workflow-catalog/#standardized">snakemake workflow catalog</a>.
Just click on the <code>Usage</code> button of any of the listed standardized workflows to get a walkthrough of how to set up this workflow for your own data analysis (for example the <a href="https://snakemake.github.io/snakemake-workflow-catalog/?usage=snakemake-workflows/dna-seq-varlociraptor"><code>dna-seq-varlociraptor</code> workflow <code>Usage</code></a>).</p>
<h2 id="snakemake-wrappers-reusable-snakemake-rules"><a class="header" href="#snakemake-wrappers-reusable-snakemake-rules">Snakemake wrappers: reusable snakemake rules</a></h2>
<p><a href="https://snakemake-wrappers.readthedocs.io">Snakemake wrappers</a> are reusable snakemake rules.
They wrap one command of a command line tool to provide a specific output from a provided input file.
You can simply copy-paste such a wrapper into your own workflow use it.
Just check out the compendium of <a href="https://snakemake-wrappers.readthedocs.io/en/stable/wrappers.html">wrappers</a> and <a href="https://snakemake-wrappers.readthedocs.io/en/stable/meta-wrappers.html">meta-wrappers</a> (copy-pastable collections of wrappers to perform analysis multiple steps).
And if the tool or command you are looking for is not yet on there, consider <a href="https://snakemake-wrappers.readthedocs.io/en/stable/contributing.html">contributing a new snakemake wrapper</a>, once you have got the tool working in your own workflow.</p>
<h2 id="snakemake-debugging-tricks"><a class="header" href="#snakemake-debugging-tricks">Snakemake debugging tricks</a></h2>
<h3 id="saving-time-on-dag-building-and-resolution"><a class="header" href="#saving-time-on-dag-building-and-resolution">Saving time on DAG building and resolution</a></h3>
<p>To quickly debug a particular rule, specify the output of that rule as the desired output of your <code>snakemake</code> run. To avoid clashes with command line argument specifications, it is best to provide the desired output file as the 1st argument right after <code>snakemake</code>, e.g.:</p>
<pre><code class="language-python">snakemake path/to/wanted_output.txt --jobs 5 --use-conda
</code></pre>
<h3 id="language-specific-debugging"><a class="header" href="#language-specific-debugging">language-specific debugging</a></h3>
<h4 id="r-1"><a class="header" href="#r-1">R</a></h4>
<h5 id="logging"><a class="header" href="#logging">logging</a></h5>
<p>Standard code we use for redirecting any <code>stderr</code> output to the <code>log:</code> file specified in the rule calling the <code>R</code> script:</p>
<pre><code class="language-r">log &lt;- file(snakemake@log[[1]], open=&quot;wt&quot;)
sink(log)
sink(log, type=&quot;message&quot;)
</code></pre>
<h5 id="interactive-debugging"><a class="header" href="#interactive-debugging">interactive debugging</a></h5>
<p>You can save the entire current state of a workspace in <code>R</code>, so you can insert this right before some code triggers an error:</p>
<pre><code class="language-r">save.image(file = &quot;my_dump.RData&quot;)
</code></pre>
<p>In an interactive R session, you can then load this workspace and interactively explore / debug what’s going on:</p>
<pre><code class="language-r">load(&quot;my_dump.RData&quot;)
</code></pre>
<h4 id="python-1"><a class="header" href="#python-1">python</a></h4>
<h5 id="logging-1"><a class="header" href="#logging-1">logging</a></h5>
<p>Standard code we use for redirecting any <code>stderr</code> output to the <code>log:</code> file specified in the rule calling the python script:</p>
<pre><code class="language-python">import sys
sys.stderr = open(snakemake.log[0], &quot;w&quot;)
</code></pre>
<h5 id="interactive-debugging-1"><a class="header" href="#interactive-debugging-1">interactive debugging</a></h5>
<p>Inserting the following into a python script executed in a snakemake rule will throw you into an interactive python debugging shell where you have access to the current state, including <code>locals()</code> and <code>globals()</code> variables:</p>
<pre><code class="language-python">import pdb; pdb.set_trace()
</code></pre>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/workflows/snakemake.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="working-on-servers"><a class="header" href="#working-on-servers">Working on servers</a></h1>
<div style="break-before: page; page-break-before: always;"></div><h1 id="persistent-screen-sessions"><a class="header" href="#persistent-screen-sessions">persistent screen sessions</a></h1>
<p><a href="https://www.gnu.org/software/screen/manual/screen.html"><code>screen</code></a> is a tool that allows you to set up persistent <code>bash</code> sessions on a server (or another computer) that can continue to run when you log off / disconnect from them and that can have multiple windows (shell session) open at once.</p>
<h2 id="getting-started"><a class="header" href="#getting-started">Getting Started</a></h2>
<p>The bare essentials are described in <a href="https://linuxize.com/post/how-to-use-linux-screen/">this <strong>How To Use Linux Screen</strong> guide</a>.
And this <a href="https://aperiodic.net/screen/quick_reference"><strong><code>screen</code> Quick Reference</strong></a> gives you the most important command line arguments and keystrokes.
For a more detailed intro, check out the <a href="https://www.gnu.org/software/screen/manual/screen.html#Getting-Started"><strong>Getting Started</strong> section of <code>screen</code>'s online documentation</a>.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/servers/screen.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="ssh"><a class="header" href="#ssh">ssh</a></h1>
<p><a href="https://www.ssh.com/academy/ssh"><code>ssh</code></a> allows for secure remote access to servers or other computers over insecure networks (including the internet).</p>
<h2 id="getting-started-with-ssh"><a class="header" href="#getting-started-with-ssh">getting started with ssh</a></h2>
<p>While there is <a href="https://www.ssh.com/academy/ssh/command">extensive documentation on <code>ssh</code> commands</a> and related commands such as <a href="https://www.ssh.com/academy/ssh/scp"><code>scp</code> (for copying files using the ssh protocol)</a>, the following Wiki page will probably get you started pretty quickly:</p>
<ul>
<li><a href="https://www.wikihow.com/Use-SSH">WikiHow page <strong>How to Use SSH</strong></a></li>
</ul>
<h2 id="ssh-keys"><a class="header" href="#ssh-keys">ssh keys</a></h2>
<p>ssh key files are a great way to set up safe ssh connections, and if properly set up with your operating system keychain, you can save on a lot of password typing.
There's a <a href="https://www.ssh.com/academy/ssh/copy-id">great guide for generating and setting up ssh keys</a> that you can quickly work through step by step.
No need to remember any of this by heart, just look it up when you need it.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/servers/ssh.md">Edit this page on GitHub.</a></footer><div style="break-before: page; page-break-before: always;"></div><h1 id="tmux"><a class="header" href="#tmux">TMUX</a></h1>
<p><a href="https://github.com/tmux/tmux/wiki">Tmux</a> is a 'terminal multiplexer' for running commands and workflows in persistent terminal sessions that you can log out of.
Thus, commands (and terminal session histories) will be kept alive even when logged out, and you can easily re-attach to a session that is still running. </p>
<h2 id="essential-instructions"><a class="header" href="#essential-instructions">Essential instructions</a></h2>
<ol>
<li>Create a new session with a name:</li>
</ol>
<pre><code>tmux new -s test
</code></pre>
<ol start="2">
<li>Detach the session that is created with <code>Ctrl+B</code> and release <code>Ctrl</code> and <code>B</code>, then quickly <code>D</code>, so it's basically <code>Ctrl+B</code> and <code>D</code>. The release of <code>Ctrl</code> and <code>B</code> is important to not <em>kill</em> the session instead of <em>detaching</em> (yes, it happens). The message should indicate: </li>
</ol>
<pre><code>[detached (from session test)]
</code></pre>
<ol start="3">
<li>List the session that is created:</li>
</ol>
<pre><code>tmux ls
</code></pre>
<ol start="4">
<li>Then, to be able to reach the newly created session back again:</li>
</ol>
<pre><code>tmux attach -t test
</code></pre>
<h2 id="other-shortcuts-that-can-be-useful"><a class="header" href="#other-shortcuts-that-can-be-useful">Other shortcuts that can be useful</a></h2>
<pre><code>CTRL - B - [ //scroll mode
q //exit scroll mode
</code></pre>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/servers/tmux.md">Edit this page on GitHub.</a></footer>
                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->


                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">

            </nav>

        </div>




        <script>
            window.playground_copyable = true;
        </script>


        <script src="elasticlunr.min.js"></script>
        <script src="mark.min.js"></script>
        <script src="searcher.js"></script>

        <script src="clipboard.min.js"></script>
        <script src="highlight.js"></script>
        <script src="book.js"></script>

        <!-- Custom JS scripts -->

        <script>
        window.addEventListener('load', function() {
            window.setTimeout(window.print, 100);
        });
        </script>

    </div>
    </body>
</html>
