<!DOCTYPE HTML>
<html lang="en" class="light" dir="ltr">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>liftover between references - A collection of resources for data scientists (not only) in bioinformatics</title>


        <!-- Custom HTML head -->
        
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff">

        <link rel="icon" href="../favicon.svg">
        <link rel="shortcut icon" href="../favicon.png">
        <link rel="stylesheet" href="../css/variables.css">
        <link rel="stylesheet" href="../css/general.css">
        <link rel="stylesheet" href="../css/chrome.css">
        <link rel="stylesheet" href="../css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="../FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="../fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="../highlight.css">
        <link rel="stylesheet" href="../tomorrow-night.css">
        <link rel="stylesheet" href="../ayu-highlight.css">

        <!-- Custom theme stylesheets -->

    </head>
    <body class="sidebar-visible no-js">
    <div id="body-container">
        <!-- Provide site root to javascript -->
        <script>
            var path_to_root = "../";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('light')
            html.classList.add(theme);
            var body = document.querySelector('body');
            body.classList.remove('no-js')
            body.classList.add('js');
        </script>

        <input type="checkbox" id="sidebar-toggle-anchor" class="hidden">

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            var body = document.querySelector('body');
            var sidebar = null;
            var sidebar_toggle = document.getElementById("sidebar-toggle-anchor");
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            sidebar_toggle.checked = sidebar === 'visible';
            body.classList.remove('sidebar-visible');
            body.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded affix "><a href="../index.html">Introduction</a></li><li class="chapter-item expanded "><a href="../bioinformatics/index.html"><strong aria-hidden="true">1.</strong> Bioinformatics</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../bioinformatics/bioconda.html"><strong aria-hidden="true">1.1.</strong> bioconda</a></li><li class="chapter-item expanded "><a href="../bioinformatics/bioconductor.html"><strong aria-hidden="true">1.2.</strong> bioconductor</a></li></ol></li><li class="chapter-item expanded "><a href="../coding_tools/index.html"><strong aria-hidden="true">2.</strong> Coding tools</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../coding_tools/conventions.html"><strong aria-hidden="true">2.1.</strong> coding conventions</a></li><li class="chapter-item expanded "><a href="../coding_tools/ide.html"><strong aria-hidden="true">2.2.</strong> integrated development environments (IDEs)</a></li><li class="chapter-item expanded "><a href="../coding_tools/licensing.html"><strong aria-hidden="true">2.3.</strong> licensing</a></li><li class="chapter-item expanded "><a href="../coding_tools/markdown.html"><strong aria-hidden="true">2.4.</strong> markdown syntax</a></li><li class="chapter-item expanded "><a href="../coding_tools/regular_expressions.html"><strong aria-hidden="true">2.5.</strong> regular expressions</a></li><li class="chapter-item expanded "><a href="../coding_tools/user_experience.html"><strong aria-hidden="true">2.6.</strong> user experience</a></li><li class="chapter-item expanded "><a href="../coding_tools/version_control.html"><strong aria-hidden="true">2.7.</strong> version control</a></li></ol></li><li class="chapter-item expanded "><a href="../command_line/index.html"><strong aria-hidden="true">3.</strong> Command line</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../command_line/bash.html"><strong aria-hidden="true">3.1.</strong> Unix shell / linux command line</a></li><li class="chapter-item expanded "><a href="../command_line/conda.html"><strong aria-hidden="true">3.2.</strong> conda (software installation)</a></li><li class="chapter-item expanded "><a href="../command_line/linux_tools.html"><strong aria-hidden="true">3.3.</strong> linux tools</a></li></ol></li><li class="chapter-item expanded "><a href="../data_analyses/index.html"><strong aria-hidden="true">4.</strong> Data analyses</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../data_analyses/differential_expression.html"><strong aria-hidden="true">4.1.</strong> differential expression analysis</a></li><li class="chapter-item expanded "><a href="../data_analyses/single_cell_analysis.html"><strong aria-hidden="true">4.2.</strong> single cell analysis</a></li></ol></li><li class="chapter-item expanded "><a href="../data_downloads/index.html"><strong aria-hidden="true">5.</strong> Data downloads / reanalysis</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../data_downloads/sra_ena_ncbi_ensemble.html"><strong aria-hidden="true">5.1.</strong> SRA, ENA, NCBI, Ensembl</a></li><li class="chapter-item expanded "><a href="../data_downloads/tcga_target_datasets.html"><strong aria-hidden="true">5.2.</strong> TCGA, TARGET, GTEx</a></li></ol></li><li class="chapter-item expanded "><a href="../data_formats/index.html"><strong aria-hidden="true">6.</strong> Data formats and tools</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../data_formats/fastq.html"><strong aria-hidden="true">6.1.</strong> fastq: sequencing data</a></li><li class="chapter-item expanded "><a href="../data_formats/tsv.html"><strong aria-hidden="true">6.2.</strong> tab-/comma-separated values</a></li><li class="chapter-item expanded "><a href="../data_formats/yaml.html"><strong aria-hidden="true">6.3.</strong> YAML syntax</a></li></ol></li><li class="chapter-item expanded "><a href="../data_presentation/index.html"><strong aria-hidden="true">7.</strong> Data presentation</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../data_presentation/color_palettes.html"><strong aria-hidden="true">7.1.</strong> color palettes</a></li><li class="chapter-item expanded "><a href="../data_presentation/grammar_of_graphics.html"><strong aria-hidden="true">7.2.</strong> grammar of graphics</a></li><li class="chapter-item expanded "><a href="../coding_tools/interactive_reporting.html"><strong aria-hidden="true">7.3.</strong> interactive reporting</a></li></ol></li><li class="chapter-item expanded "><a href="../languages/index.html"><strong aria-hidden="true">8.</strong> Programming Languages</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../languages/python.html"><strong aria-hidden="true">8.1.</strong> python</a></li><li class="chapter-item expanded "><a href="../languages/r.html"><strong aria-hidden="true">8.2.</strong> R</a></li><li class="chapter-item expanded "><a href="../languages/rust.html"><strong aria-hidden="true">8.3.</strong> Rust</a></li></ol></li><li class="chapter-item expanded "><a href="../reference_data/index.html"><strong aria-hidden="true">9.</strong> Reference data</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../reference_data/liftover.html" class="active"><strong aria-hidden="true">9.1.</strong> liftover between references</a></li></ol></li><li class="chapter-item expanded "><a href="../statistics/index.html"><strong aria-hidden="true">10.</strong> Statistics</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../statistics/bayesian.html"><strong aria-hidden="true">10.1.</strong> bayesian statistics (priors, likelihoods and posterior distributions)</a></li><li class="chapter-item expanded "><a href="../statistics/frequentist.html"><strong aria-hidden="true">10.2.</strong> frequentist statistics (hypothesis tests, p-values, etc.)</a></li></ol></li><li class="chapter-item expanded "><a href="../workflows/index.html"><strong aria-hidden="true">11.</strong> Workflow management</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../workflows/nextflow.html"><strong aria-hidden="true">11.1.</strong> Nextflow</a></li><li class="chapter-item expanded "><a href="../workflows/snakemake.html"><strong aria-hidden="true">11.2.</strong> Snakemake</a></li></ol></li><li class="chapter-item expanded "><a href="../servers/index.html"><strong aria-hidden="true">12.</strong> Working on servers</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../servers/persistent_shell.html"><strong aria-hidden="true">12.1.</strong> persistent shell sessions (screen, tmux)</a></li><li class="chapter-item expanded "><a href="../servers/ssh.html"><strong aria-hidden="true">12.2.</strong> ssh login</a></li><li class="chapter-item expanded "><a href="../servers/data_transfer.html"><strong aria-hidden="true">12.3.</strong> data transfer</a></li></ol></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle">
                <div class="sidebar-resize-indicator"></div>
            </div>
        </nav>

        <!-- Track and set sidebar scroll position -->
        <script>
            var sidebarScrollbox = document.querySelector('#sidebar .sidebar-scrollbox');
            sidebarScrollbox.addEventListener('click', function(e) {
                if (e.target.tagName === 'A') {
                    sessionStorage.setItem('sidebar-scroll', sidebarScrollbox.scrollTop);
                }
            }, { passive: true });
            var sidebarScrollTop = sessionStorage.getItem('sidebar-scroll');
            sessionStorage.removeItem('sidebar-scroll');
            if (sidebarScrollTop) {
                // preserve sidebar scroll position when navigating via links within sidebar
                sidebarScrollbox.scrollTop = sidebarScrollTop;
            } else {
                // scroll sidebar to current active section when navigating via "next/previous chapter" buttons
                var activeSection = document.querySelector('#sidebar .active');
                if (activeSection) {
                    activeSection.scrollIntoView({ block: 'center' });
                }
            }
        </script>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky">
                    <div class="left-buttons">
                        <label id="sidebar-toggle" class="icon-button" for="sidebar-toggle-anchor" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </label>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">A collection of resources for data scientists (not only) in bioinformatics</h1>

                    <div class="right-buttons">
                        <a href="../print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>
                        <a href="https://github.com/koesterlab/data-science-for-bioinfo" title="Git repository" aria-label="Git repository">
                            <i id="git-repository-button" class="fa fa-github"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="liftover-between-reference-genomes"><a class="header" href="#liftover-between-reference-genomes">Liftover between reference genomes</a></h1>
<p>Tasks liks read mapping and variant calling are executed with respect to a reference genome build or version.
For <em>Homo sapiens</em> there are currently two commonly used builds: GRCh37 (hg19) and GRCh38 (hg38).
Sometimes you need to convert (liftover) coordinates from one reference build to another, for example to be able to use coordinate-specific annotations or to correctly compare coordinate-specific results file.
Or you might want to do something funkier, like lifting over coordinates between different species.
Here, we collect tools and resources for different coordinate-specific formats and how to use them.</p>
<h2 id="chain-files"><a class="header" href="#chain-files">chain files</a></h2>
<blockquote>
<p>A <a href="https://genome.ucsc.edu/goldenPath/help/chain.html">chain file</a> describes a pairwise alignment between two reference assemblies. (From the <a href="https://crossmap.sourceforge.net/#chain-file">CrossMap docs</a>)
All the tools mentioned below use a chain file for the liftover process, so will need one for your pair of reference genomes.</p>
</blockquote>
<h3 id="downloading-chain-files"><a class="header" href="#downloading-chain-files">downloading chain files</a></h3>
<p>The <a href="https://crossmap.sourceforge.net/#chain-file"><code>CrossMap</code> documentation on chain files</a> has a good link collection with chain files available from Ensembl and UCSC.
The <a href="https://doi.org/10.1093/bioinformatics/btae038"><code>bcftools +liftover</code> paper</a> suggests, that the UCSC chain files are more comprehensive than the Ensembl chain files:</p>
<blockquote>
<p>For the liftover from GRCh37 to GRCh38, we did not use the Ensembl chain file (GRCh37_to_GRCh38.chain.gz) generated from Ensembl assembly mappings (http://github.com/Ensembl/ensembl/) as this resulted in a much higher rate of variants dropped compared with using the UCSC chain file.</p>
</blockquote>
<h3 id="creating-chain-files"><a class="header" href="#creating-chain-files">creating chain files</a></h3>
<p>These tools are untested, but we document them here to try out in the future.
Please report back to this knowledge base on their usage if you come to try them out:</p>
<ul>
<li><a href="https://nf-lo.readthedocs.io/en/latest/index.html">nextflow LiftOver workflow</a>: automatically downloads NCBI or iGenomes references and implements several aligners to create chain files within or between species</li>
<li><a href="https://github.com/informationsea/transanno">transanno</a> can do VCF /GFF3 / GTF file liftover (but bcftools liftover is supposedly better at this), but also provides the <a href="https://github.com/informationsea/transanno?tab=readme-ov-file#create-chain-file-from-minimap2-result"><code>transanno minimap2-to-chain</code> command for creating chain files from minimap2 PAF alignments</a></li>
</ul>
<h2 id="tools"><a class="header" href="#tools">tools</a></h2>
<h3 id="vcf-liftover"><a class="header" href="#vcf-liftover">VCF liftover</a></h3>
<h4 id="picard-liftovervcf"><a class="header" href="#picard-liftovervcf">picard <code>LiftoverVcf</code></a></h4>
<p>For VCF files, Picard offers the command <a href="http://broadinstitute.github.io/picard/command-line-overview.html#LiftoverVcf"><code>LiftoverVcf</code></a>.</p>
<p><strong>Tested usage for a referene fasta and a callset vcf:</strong></p>
<pre><code>mamba create -n picard picard
mamba activate picard
picard CreateSequenceDictionary  -R {input.reference} -O {input.reference}.dict
picard LiftoverVcf  -I {input.callset}  -O {output} --CHAIN {input.liftover_chain} --REJECT {output}_rejected_variants.vcf -R {input.reference}
</code></pre>
<h3 id="bcftools-liftover-plugin"><a class="header" href="#bcftools-liftover-plugin"><code>bcftools +liftover</code> plugin</a></h3>
<p>This looks like a very thorough tool to lift over indels (in addition to SNVs, which all tools are doing OK), judging from the great <a href="https://doi.org/10.1093/bioinformatics/btae038">paper introducing <code>bcftools +liftover</code> plugin</a>.
While installation is not easily automated, yet, it is definitely worth trying out <a href="https://github.com/freeseek/score/issues/4">once it is integrated as a plugin into bcftools itself</a>.
Then, it should be <a href="https://github.com/freeseek/score/tree/master?tab=readme-ov-file#liftover-vcfs">pretty straightforward to conduct a liftover</a>.</p>
<h3 id="other-format-liftover"><a class="header" href="#other-format-liftover">other format liftover</a></h3>
<h4 id="crossmap"><a class="header" href="#crossmap">CrossMap</a></h4>
<p><a href="https://crossmap.sourceforge.net/">CrossMap</a> allows for a wider range of formats to be lifted over, including <strong>SAM/BAM/CRAM, VCF, BED, GFF/GTF, MAF, BigWig and Wiggle format</strong>.
It is <a href="https://bioconda.github.io/recipes/crossmap/README.html">available for installation via <code>bioconda</code></a> and has <a href="https://crossmap.sourceforge.net/#usage">very clear usage instructions</a>, with detailed information for the different formats it supports.
However, analyses by the authors of <code>bcftools +liftover</code> and <code>transanno</code> suggest, that the VCF/BCF-specific tools perform better on this data type.</p>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/reference_data/liftover.md">Edit this page on GitHub.</a></footer>
                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                            <a rel="prev" href="../reference_data/index.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <i class="fa fa-angle-left"></i>
                            </a>

                            <a rel="next prefetch" href="../statistics/index.html" class="mobile-nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                                <i class="fa fa-angle-right"></i>
                            </a>

                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                    <a rel="prev" href="../reference_data/index.html" class="nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                        <i class="fa fa-angle-left"></i>
                    </a>

                    <a rel="next prefetch" href="../statistics/index.html" class="nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                        <i class="fa fa-angle-right"></i>
                    </a>
            </nav>

        </div>




        <script>
            window.playground_copyable = true;
        </script>


        <script src="../elasticlunr.min.js"></script>
        <script src="../mark.min.js"></script>
        <script src="../searcher.js"></script>

        <script src="../clipboard.min.js"></script>
        <script src="../highlight.js"></script>
        <script src="../book.js"></script>

        <!-- Custom JS scripts -->


    </div>
    </body>
</html>
