<!DOCTYPE HTML>
<html lang="en" class="light" dir="ltr">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>differential expression analysis - A collection of resources for data scientists (not only) in bioinformatics</title>


        <!-- Custom HTML head -->
        
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff">

        <link rel="icon" href="../favicon.svg">
        <link rel="shortcut icon" href="../favicon.png">
        <link rel="stylesheet" href="../css/variables.css">
        <link rel="stylesheet" href="../css/general.css">
        <link rel="stylesheet" href="../css/chrome.css">
        <link rel="stylesheet" href="../css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="../FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="../fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="../highlight.css">
        <link rel="stylesheet" href="../tomorrow-night.css">
        <link rel="stylesheet" href="../ayu-highlight.css">

        <!-- Custom theme stylesheets -->

    </head>
    <body class="sidebar-visible no-js">
    <div id="body-container">
        <!-- Provide site root to javascript -->
        <script>
            var path_to_root = "../";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('light')
            html.classList.add(theme);
            var body = document.querySelector('body');
            body.classList.remove('no-js')
            body.classList.add('js');
        </script>

        <input type="checkbox" id="sidebar-toggle-anchor" class="hidden">

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            var body = document.querySelector('body');
            var sidebar = null;
            var sidebar_toggle = document.getElementById("sidebar-toggle-anchor");
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            sidebar_toggle.checked = sidebar === 'visible';
            body.classList.remove('sidebar-visible');
            body.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded affix "><a href="../index.html">Introduction</a></li><li class="chapter-item expanded "><a href="../bioinformatics/index.html"><strong aria-hidden="true">1.</strong> Bioinformatics</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../bioinformatics/bioconda.html"><strong aria-hidden="true">1.1.</strong> bioconda</a></li><li class="chapter-item expanded "><a href="../bioinformatics/bioconductor.html"><strong aria-hidden="true">1.2.</strong> bioconductor</a></li></ol></li><li class="chapter-item expanded "><a href="../coding_tools/index.html"><strong aria-hidden="true">2.</strong> Coding tools</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../coding_tools/conventions.html"><strong aria-hidden="true">2.1.</strong> coding conventions</a></li><li class="chapter-item expanded "><a href="../coding_tools/ide.html"><strong aria-hidden="true">2.2.</strong> integrated development environments (IDEs)</a></li><li class="chapter-item expanded "><a href="../coding_tools/licensing.html"><strong aria-hidden="true">2.3.</strong> licensing</a></li><li class="chapter-item expanded "><a href="../coding_tools/markdown.html"><strong aria-hidden="true">2.4.</strong> markdown syntax</a></li><li class="chapter-item expanded "><a href="../coding_tools/regular_expressions.html"><strong aria-hidden="true">2.5.</strong> regular expressions</a></li><li class="chapter-item expanded "><a href="../coding_tools/user_experience.html"><strong aria-hidden="true">2.6.</strong> user experience</a></li><li class="chapter-item expanded "><a href="../coding_tools/version_control.html"><strong aria-hidden="true">2.7.</strong> version control</a></li></ol></li><li class="chapter-item expanded "><a href="../command_line/index.html"><strong aria-hidden="true">3.</strong> Command line</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../command_line/bash.html"><strong aria-hidden="true">3.1.</strong> Unix shell / linux command line</a></li><li class="chapter-item expanded "><a href="../command_line/conda.html"><strong aria-hidden="true">3.2.</strong> conda (software installation)</a></li><li class="chapter-item expanded "><a href="../command_line/linux_tools.html"><strong aria-hidden="true">3.3.</strong> linux tools</a></li></ol></li><li class="chapter-item expanded "><a href="../data_analyses/index.html"><strong aria-hidden="true">4.</strong> Data analyses</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../data_analyses/differential_expression.html" class="active"><strong aria-hidden="true">4.1.</strong> differential expression analysis</a></li></ol></li><li class="chapter-item expanded "><a href="../data_downloads/index.html"><strong aria-hidden="true">5.</strong> Data downloads / reanalysis</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../data_downloads/sra_ena_ncbi_ensemble.html"><strong aria-hidden="true">5.1.</strong> SRA, ENA, NCBI, Ensembl</a></li><li class="chapter-item expanded "><a href="../data_downloads/tcga_target_datasets.html"><strong aria-hidden="true">5.2.</strong> TCGA, TARGET, GTEx</a></li></ol></li><li class="chapter-item expanded "><a href="../data_formats/index.html"><strong aria-hidden="true">6.</strong> Data formats and tools</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../data_formats/fastq.html"><strong aria-hidden="true">6.1.</strong> fastq: sequencing data</a></li><li class="chapter-item expanded "><a href="../data_formats/tsv.html"><strong aria-hidden="true">6.2.</strong> tab-/comma-separated values</a></li><li class="chapter-item expanded "><a href="../data_formats/yaml.html"><strong aria-hidden="true">6.3.</strong> YAML syntax</a></li></ol></li><li class="chapter-item expanded "><a href="../data_presentation/index.html"><strong aria-hidden="true">7.</strong> Data presentation</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../data_presentation/color_palettes.html"><strong aria-hidden="true">7.1.</strong> color palettes</a></li><li class="chapter-item expanded "><a href="../data_presentation/grammar_of_graphics.html"><strong aria-hidden="true">7.2.</strong> grammar of graphics</a></li><li class="chapter-item expanded "><a href="../coding_tools/interactive_reporting.html"><strong aria-hidden="true">7.3.</strong> interactive reporting</a></li></ol></li><li class="chapter-item expanded "><a href="../languages/index.html"><strong aria-hidden="true">8.</strong> Programming Languages</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../languages/python.html"><strong aria-hidden="true">8.1.</strong> python</a></li><li class="chapter-item expanded "><a href="../languages/r.html"><strong aria-hidden="true">8.2.</strong> R</a></li><li class="chapter-item expanded "><a href="../languages/rust.html"><strong aria-hidden="true">8.3.</strong> Rust</a></li></ol></li><li class="chapter-item expanded "><a href="../reference_data/index.html"><strong aria-hidden="true">9.</strong> Reference data</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../reference_data/liftover.html"><strong aria-hidden="true">9.1.</strong> liftover between references</a></li></ol></li><li class="chapter-item expanded "><a href="../statistics/index.html"><strong aria-hidden="true">10.</strong> Statistics</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../statistics/bayesian.html"><strong aria-hidden="true">10.1.</strong> bayesian statistics (priors, likelihoods and posterior distributions)</a></li><li class="chapter-item expanded "><a href="../statistics/frequentist.html"><strong aria-hidden="true">10.2.</strong> frequentist statistics (hypothesis tests, p-values, etc.)</a></li></ol></li><li class="chapter-item expanded "><a href="../workflows/index.html"><strong aria-hidden="true">11.</strong> Workflow management</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../workflows/nextflow.html"><strong aria-hidden="true">11.1.</strong> Nextflow</a></li><li class="chapter-item expanded "><a href="../workflows/snakemake.html"><strong aria-hidden="true">11.2.</strong> Snakemake</a></li></ol></li><li class="chapter-item expanded "><a href="../servers/index.html"><strong aria-hidden="true">12.</strong> Working on servers</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../servers/persistent_shell.html"><strong aria-hidden="true">12.1.</strong> persistent shell sessions (screen, tmux)</a></li><li class="chapter-item expanded "><a href="../servers/ssh.html"><strong aria-hidden="true">12.2.</strong> ssh login</a></li><li class="chapter-item expanded "><a href="../servers/data_transfer.html"><strong aria-hidden="true">12.3.</strong> data transfer</a></li></ol></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle">
                <div class="sidebar-resize-indicator"></div>
            </div>
        </nav>

        <!-- Track and set sidebar scroll position -->
        <script>
            var sidebarScrollbox = document.querySelector('#sidebar .sidebar-scrollbox');
            sidebarScrollbox.addEventListener('click', function(e) {
                if (e.target.tagName === 'A') {
                    sessionStorage.setItem('sidebar-scroll', sidebarScrollbox.scrollTop);
                }
            }, { passive: true });
            var sidebarScrollTop = sessionStorage.getItem('sidebar-scroll');
            sessionStorage.removeItem('sidebar-scroll');
            if (sidebarScrollTop) {
                // preserve sidebar scroll position when navigating via links within sidebar
                sidebarScrollbox.scrollTop = sidebarScrollTop;
            } else {
                // scroll sidebar to current active section when navigating via "next/previous chapter" buttons
                var activeSection = document.querySelector('#sidebar .active');
                if (activeSection) {
                    activeSection.scrollIntoView({ block: 'center' });
                }
            }
        </script>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky">
                    <div class="left-buttons">
                        <label id="sidebar-toggle" class="icon-button" for="sidebar-toggle-anchor" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </label>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">A collection of resources for data scientists (not only) in bioinformatics</h1>

                    <div class="right-buttons">
                        <a href="../print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>
                        <a href="https://github.com/koesterlab/data-science-for-bioinfo" title="Git repository" aria-label="Git repository">
                            <i id="git-repository-button" class="fa fa-github"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="differential-expression-analysis"><a class="header" href="#differential-expression-analysis">differential expression analysis</a></h1>
<p>A classic case of bioinformatic data analysis, is to check whether expression levels of individual transcripts differ between two (or more) groups of samples.
These can be a treatment group compared to a control group, different time points, or groups differing according to any other annotation variable from their metadata.
Up to a certain point, these analyses differ depending on the raw data you start from (for example expression microarrays, full transcriptome RNA-seq data, or specialized RNA sequencing methods).
Thus, we have <a href="#data-types">one section per data type</a> here, and include linkouts to standardized (snakemake) workflows wherever available, which will make it much quicker for you to get started with your analysis, and much easier to update and/or redo it later.
However, after the initial data processing, differential expression analysis is usually performed by the same (statistical) methods.
Thus, we have <a href="#differential-expression-analysis-tools">another section dedicated to differential expression analysis tools</a>.
Finally, transcripts (or genes) that have been found to be differentially expressed, can be analyzed further---for example via gene set or enrichment or pathway perturbation analysis.
We list some tools for these tasks in the <a href="#analysis-of-differentially-expressed-transcripts">section <code>analysis of differentially expressed transcripts</code></a></p>
<h2 id="data-generation--experimental-design"><a class="header" href="#data-generation--experimental-design">data generation / experimental design</a></h2>
<p>A good recap of available research and recommendations is given in the <a href="https://www.nature.com/articles/s41576-019-0150-2#Sec13">section <code>Designing better RNA-seq experiments</code></a> of this review paper:</p>
<p><a href="https://doi.org/10.1038/s41576-019-0150-2">Stark, R., Grzelak, M. &amp; Hadfield, J. <strong>RNA sequencing: the teenage years.</strong> <em>Nat Rev Genet</em> 20, 631–656 (2019). https://doi.org/10.1038/s41576-019-0150-2</a></p>
<h3 id="number-of-biological-replicates"><a class="header" href="#number-of-biological-replicates">number of biological replicates</a></h3>
<p>You should <strong>always include biological replicates</strong> in any RNAseq analysis.
Differential expression analysis tools will either not work at all without biological or the result will not be very useful.
As <a href="https://support.bioconductor.org/p/95714/#95717">Michael Love, the developer of DESeq2 stated</a>, without biological replicates "you have no idea <code>[of]</code> the degree of within-group variability, so you can't perform statistical inference (determining if the difference you see is more than you would typically see among replicates within one condition)".</p>
<p>How many biological replicates to include, depends on how sensitive your analysis needs to be and how heterogeneous you expect your replicates to be within sample groups.
Anything below 3 samples is generally discouraged, as <a href="https://doi.org/10.1261%2Frna.053959.115">a dedicated study on biological replicate numbers that had low within-group variability/heterogeneity</a> could only detect about 85% of expected differentially expressed genes with a fold change of 2 or higher.
Thus, any more sublte fold changes will most likely be missed and the same study required at least 6 biological replicates to detect about 80% of expected differentially expressed genes with a fold change of 1.2 or above.</p>
<p>The general rules here are:</p>
<ol>
<li>If you want to detect differentially expressed genes with smaller effect sizes (lower fold changes), you will need more biological replicates.</li>
<li>The more heterogeneity/variability you expect within a group of biological replicates, the more biological replicates you need.
For example, in vitro experiments with biological replicates from the same cell line should allow for fewer biological replicates, while studies where samples within a group come from different patients will usually require more biological replicates per group.</li>
</ol>
<h3 id="sequencing-depth"><a class="header" href="#sequencing-depth">sequencing depth</a></h3>
<p>Generally, sequencing depth is not as important a factor as the number of biological replicates.
The above-cited review article concludes that for eukaryotic genomes, around 10 - 30 million reads per sample give good results.</p>
<h3 id="sequencing-length"><a class="header" href="#sequencing-length">sequencing length</a></h3>
<p>According to the above-cited review article, if you only want to do general differential expression analysis, single-end reads with a length of at least 50 base pairs should suffice.
If you are also interested in the expression of different transcript isoforms, the detection of fusion gene transcripts or <em>de novo</em> transcript assembly, longer paired-end reads will improve the sensitivity of your analyses.</p>
<h2 id="data-types"><a class="header" href="#data-types">data types</a></h2>
<h3 id="rna-seq-data"><a class="header" href="#rna-seq-data">RNA-seq data</a></h3>
<p>TODO: provide good self-contained tutorial for working with RNA-seq data</p>
<p>There are several standardized snakemake workflows, the following ones are maintained by the Kösterlab and associated people:</p>
<ul>
<li><a href="https://snakemake.github.io/snakemake-workflow-catalog/?usage=snakemake-workflows/rna-seq-kallisto-sleuth">a workflow using <code>kallisto</code> for quantifying transcript expression and <code>sleuth</code> (see below)</a> for determining differentially expressed transcripts</li>
<li><a href="https://snakemake.github.io/snakemake-workflow-catalog/?usage=snakemake-workflows/rna-seq-star-deseq2">a workflow using <code>Star</code> to align reads to the transcriptome and <code>DESeq2</code> (see below)</a> for determining differentially expressed transcripts</li>
</ul>
<h3 id="microarray-data"><a class="header" href="#microarray-data">microarray data</a></h3>
<p>While fewer and fewer new microarray data sets are being generated, a lot of existing microarray data sets can be re-analyzed for new research questions.
If you want to get started with such analyses, there is a great hands-on tutorial by Greg Tucker-Kellogg: <a href="https://gtk-teaching.github.io/Microarrays-R/">Introduction to gene expression microarray analysis in R and Bioconductor</a>.
It includes an introduction to Bioconductor, works on publicly available data and is completely self-contained.
You can work through it at your own pace and all it requires is some basic R knowledge---for the latter, see our section on <a href="../languages/r.html#learning-programming-with-r">learning R</a>.</p>
<p>Once you are familiar with the data type and the analysis, there's a <a href="https://snakemake.github.io/snakemake-workflow-catalog/?usage=zifornd/arrays">standardized snakemake workflow for microarray analysis</a> by James Ashmore.
While we haven't used this workflow ourselves (but might in the future, we'll try to update the info here if we do), this workflow is:</p>
<ul>
<li>standardized, so:
<ul>
<li>the above link should take you through all of the setup steps you will need to run this workflow on your own data</li>
<li>it is well-structured, so modifying it to your needs should be relatively easy</li>
</ul>
</li>
<li>based on a <a href="https://doi.org/10.12688/f1000research.8967.2">thoroughly peer reviewed walk-through manuscript on microarray analysis</a></li>
</ul>
<p>This manuscript also gives good and detailed recommendations for microarray analysis in general, so you can also consult it if you are just looking for help on some certain analysis steps.</p>
<h2 id="differential-expression-analysis-tools"><a class="header" href="#differential-expression-analysis-tools">differential expression analysis tools</a></h2>
<h3 id="sleuth"><a class="header" href="#sleuth">sleuth</a></h3>
<p>TODO: add some info
<a href="https://pachterlab.github.io/sleuth/walkthroughs"><code>sleuth</code> walkthroughs for different analyses types</a></p>
<h3 id="deseq2"><a class="header" href="#deseq2">DeSEQ2</a></h3>
<p>TODO: add some info
<a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">extensive HTML documentation page for <code>DESeq2</code></a>
<a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">main bioconductor package page for <code>DESeq2</code></a></p>
<h3 id="edger"><a class="header" href="#edger">edgeR</a></h3>
<p>TODO: add some info
<a href="https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf">extensive PDF user guide for <code>edgeR</code></a>
<a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">main bioconductor package page for <code>edgeR</code></a></p>
<h2 id="analysis-of-differentially-expressed-transcripts"><a class="header" href="#analysis-of-differentially-expressed-transcripts">analysis of differentially expressed transcripts</a></h2>
<h3 id="gene-set-enrichment"><a class="header" href="#gene-set-enrichment">gene set enrichment</a></h3>
<h3 id="pathway-enrichment"><a class="header" href="#pathway-enrichment">pathway enrichment</a></h3>
<h3 id="pathway-perturbation"><a class="header" href="#pathway-perturbation">pathway perturbation</a></h3>
<footer id="open-on-gh">Want to add or correct something? <a href="https://github.com/koesterlab/data-science-for-bioinfo/edit/main/src/data_analyses/differential_expression.md">Edit this page on GitHub.</a></footer>
                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                            <a rel="prev" href="../data_analyses/index.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <i class="fa fa-angle-left"></i>
                            </a>

                            <a rel="next prefetch" href="../data_downloads/index.html" class="mobile-nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                                <i class="fa fa-angle-right"></i>
                            </a>

                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                    <a rel="prev" href="../data_analyses/index.html" class="nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                        <i class="fa fa-angle-left"></i>
                    </a>

                    <a rel="next prefetch" href="../data_downloads/index.html" class="nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                        <i class="fa fa-angle-right"></i>
                    </a>
            </nav>

        </div>




        <script>
            window.playground_copyable = true;
        </script>


        <script src="../elasticlunr.min.js"></script>
        <script src="../mark.min.js"></script>
        <script src="../searcher.js"></script>

        <script src="../clipboard.min.js"></script>
        <script src="../highlight.js"></script>
        <script src="../book.js"></script>

        <!-- Custom JS scripts -->


    </div>
    </body>
</html>
